diff -uprN a/build_opk.sh b/build_opk.sh
--- a/build_opk.sh	1970-01-01 08:00:00.000000000 +0800
+++ b/build_opk.sh	2026-03-07 11:19:59.236318728 +0800
@@ -0,0 +1,7 @@
+#!/bin/bash
+
+make PLATFORM=bittboy
+/opt/mmiyoo/bin/arm-linux-gnueabihf-strip abbaye -o ./_opk/abbaye
+mv fonts graphics mapa music sounds ./_opk/
+mksquashfs ./_opk/* ../_opk/abbaye_funkey-s.opk -all-root -no-xattrs -noappend -no-exports
+cd ..
diff -uprN a/makefile b/makefile
--- a/makefile	2026-02-17 15:37:41.654299944 +0800
+++ b/makefile	2026-02-17 15:41:37.964301170 +0800
@@ -26,10 +26,10 @@ ifeq ($(PLATFORM),gcw0)
     CC     = mipsel-linux-gcc
     CFLAGS += -mips32 -D_GCW_ZERO
 else ifeq ($(PLATFORM),bittboy)
-    CC     = /opt/miyoo-toolchain-old/bin/arm-linux-gcc
-    CFLAGS += -D_BITTBOY -I/opt/miyoo-toolchain-old/arm-miyoo-linux-uclibcgnueabi/sysroot/usr/include/SDL
-    SDL_CFLAGS := `/opt/miyoo-toolchain-old/arm-miyoo-linux-uclibcgnueabi/sysroot/usr/bin/sdl-config --cflags`
-    SDL_LIBS   := `/opt/miyoo-toolchain-old/arm-miyoo-linux-uclibcgnueabi/sysroot/usr/bin/sdl-config --libs`
+    CC     = /opt/mmiyoo/bin/arm-linux-gnueabihf-gcc
+    CFLAGS += -D_BITTBOY -I/opt/mmiyoo/arm-buildroot-linux-gnueabihf/sysroot/usr/include/SDL
+    SDL_CFLAGS := `/opt/mmiyoo/arm-buildroot-linux-gnueabihf/sysroot/usr/bin/sdl-config --cflags`
+    SDL_LIBS   := `/opt/mmiyoo/arm-buildroot-linux-gnueabihf/sysroot/usr/bin/sdl-config --libs`
 else ifeq ($(PLATFORM),pandora)
     NAME   = abbaye-pandora
     CC     = $(PREFIX)gcc
diff -uprN a/src/main.c b/src/main.c
--- a/src/main.c	2026-02-17 15:38:26.544300177 +0800
+++ b/src/main.c	2026-02-17 15:44:43.214302131 +0800
@@ -1267,42 +1267,42 @@ void tecladoj (struct protagonista *jean
 				if ((jean->pulsa[0] == 0) && (jean->salto == 0) && (jean->agachado == 0))
 					jean->salto = 1;
 			}
-			if (evento.key.keysym.sym == SDLK_DOWN) {
+			if (evento.key.keysym.sym == SDLK_DOWN) { 
 				if ((jean->pulsa[1] == 0) && (jean->salto == 0)) {
 					jean->pulsa[1] = 1;
 					jean->agachado = 1;
 				}
 			}
-			if (evento.key.keysym.sym == SDLK_LEFT) {
+			if (evento.key.keysym.sym == SDLK_LEFT) { 
 				if (jean->pulsa[2] == 0) {
 					jean->pulsa[2] = 1;
 					jean->pulsa[3] = 0;
 				}
 			}
-			if (evento.key.keysym.sym == SDLK_RIGHT) {
+			if (evento.key.keysym.sym == SDLK_RIGHT) { 
 				if (jean->pulsa[3] == 0) {
 					jean->pulsa[3] = 1;
 					jean->pulsa[2] = 0;
 				}
 			}
-			if (evento.key.keysym.sym == SDLK_f)
+			if (evento.key.keysym.sym == SDLK_m)
 				*tecla = 6;
 			if (evento.key.keysym.sym == KEY_GRAPHICS)
 				*tecla = 9;
-	   	if (evento.key.keysym.sym == SDLK_RCTRL)
+	   	if (evento.key.keysym.sym == SDLK_ESCAPE)
       		exit(0);
 		}
 
 		if (evento.type == SDL_KEYUP) {
 			if (evento.key.keysym.sym == KEY_JUMP)
 					jean->pulsa[0] = 0;
-			if (evento.key.keysym.sym == SDLK_DOWN) {
+			if (evento.key.keysym.sym == SDLK_DOWN) { 
 					jean->pulsa[1] = 0;
 					jean->agachado = 0;
 			}
-			if (evento.key.keysym.sym == SDLK_LEFT)
+			if (evento.key.keysym.sym == SDLK_LEFT) 
 					jean->pulsa[2] = 0;
-			if (evento.key.keysym.sym == SDLK_RIGHT)
+			if (evento.key.keysym.sym == SDLK_RIGHT) 
 					jean->pulsa[3] = 0;
 		}
 #endif
@@ -1369,9 +1369,9 @@ void teclado (int *tecla, int fase) {
       		if (fase == 1)
 						*tecla = 7;
 			}
-	   	if (evento.key.keysym.sym == SDLK_RCTRL)
+	   	if (evento.key.keysym.sym == SDLK_ESCAPE)
       		exit(0);
-			if (evento.key.keysym.sym == SDLK_f)
+			if (evento.key.keysym.sym == SDLK_m)
 				*tecla = 6;
 			if (evento.key.keysym.sym == KEY_GRAPHICS)
 				*tecla = 9;
@@ -1405,7 +1405,7 @@ void tecladop (int *teclap) {
 #ifndef JOYSTICK_SUPPORT
 		if (evento.type == SDL_KEYDOWN) {
 			if ((evento.key.keysym.sym == KEY_START) ||
-					(evento.key.keysym.sym == SDLK_LEFT) || (evento.key.keysym.sym == SDLK_RIGHT))
+					(evento.key.keysym.sym == SDLK_LEFT) || (evento.key.keysym.sym == SDLK_RIGHT)) 
 				*teclap = 1;
 		}
 #endif
