diff -uprN a/build_opk.sh b/build_opk.sh
--- a/build_opk.sh	1970-01-01 08:00:00.000000000 +0800
+++ b/build_opk.sh	2026-03-08 14:15:52.761246258 +0800
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+export C_FLAGS="-std=c99 -Wall -Wextra -pedantic -O3 -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -o anarch"
+export COMPILER="/opt/mmiyoo/bin/arm-linux-gnueabihf-g++"
+export SYSROOT=`${COMPILER} --print-sysroot`
+export SDL_FLAGS=`${SYSROOT}/usr/bin/sdl-config --libs`
+export COMMAND="${COMPILER} ${C_FLAGS} main_sdl1.c ${SDL_FLAGS} -DMIYOO"
+${COMMAND}
+/opt/mmiyoo/bin/arm-linux-gnueabihf-strip anarch -o _opk/anarch
+mksquashfs _opk/* ../_opk/anarch_funkey-s.opk -all-root -no-xattrs -noappend -no-exports
