diff -uprN a/build_opk.sh b/build_opk.sh
--- a/build_opk.sh	1970-01-01 08:00:00.000000000 +0800
+++ b/build_opk.sh	2026-03-08 19:02:20.954056772 +0800
@@ -0,0 +1,6 @@
+#!/bin/bash
+mv _opk/sysevt.c src/
+make
+/opt/mmiyoo/bin/arm-linux-gnueabihf-strip xrick.dge -o _opk/xrick
+mv data.zip _opk/
+mksquashfs _opk/* ../_opk/xrick_funkey-s.opk -all-root -no-xattrs -noappend -no-exports
diff -uprN a/Makefile b/Makefile
--- a/Makefile	2026-02-16 12:09:30.906634223 +0800
+++ b/Makefile	2026-02-16 12:13:48.816635562 +0800
@@ -15,7 +15,7 @@
 # Vars
 #
 
-SDLVERSION=$(shell /opt/miyoo/arm-miyoo-linux-uclibcgnueabi/sysroot/usr/bin/sdl-config --version 2>/dev/null)
+SDLVERSION=$(shell /opt/mmiyoo/arm-buildroot-linux-gnueabihf/sysroot/usr/bin/sdl-config --version 2>/dev/null)
 ROOTDIR=
 #$(shell pwd)
 TARGET=$(shell uname -s | tr [a-z] [A-Z])
@@ -83,10 +83,10 @@ endif
 all:
 	@echo "ROOTDIR=" > Makefile.global
 	@echo "XOBJ=$(XOBJ)" >> Makefile.global
-	@echo "CFLAGS=-g -ansi -pedantic -Wall -W -O2 -I $(ROOTDIR)/include $(shell /opt/miyoo/arm-miyoo-linux-uclibcgnueabi/sysroot/usr/bin/sdl-config --cflags) $(OPTIMISE)" >> Makefile.global
-	@echo "LDFLAGS=-lz $(shell /opt/miyoo/arm-miyoo-linux-uclibcgnueabi/sysroot/usr/bin/sdl-config --libs) -fprofile-arcs" >> Makefile.global
-	@echo "CC=arm-miyoo-linux-uclibcgnueabi-gcc" >> Makefile.global
-	@echo "CPP=arm-miyoo-linux-uclibcgnueabi-gcc -E" >> Makefile.global
+	@echo "CFLAGS=-g -ansi -pedantic -Wall -W -O2 -I $(ROOTDIR)/include $(shell /opt/mmiyoo/arm-buildroot-linux-gnueabihf/sysroot/usr/bin/sdl-config --cflags) $(OPTIMISE)" >> Makefile.global
+	@echo "LDFLAGS=-lz $(shell /opt/mmiyoo/arm-buildroot-linux-gnueabihf/sysroot/usr/bin/sdl-config --libs) -fprofile-arcs" >> Makefile.global
+	@echo "CC=/opt/mmiyoo/bin/arm-linux-gnueabihf-gcc" >> Makefile.global
+	@echo "CPP=/opt/mmiyoo/bin/arm-linux-gnueabihf-gcc -E" >> Makefile.global
 	$(MAKE) -C src all
 
 clean:
