diff -uprN a/sdl2/src/video/mmiyoo/SDL_event_mmiyoo.c b/sdl2/src/video/mmiyoo/SDL_event_mmiyoo.c
--- a/sdl2/src/video/mmiyoo/SDL_event_mmiyoo.c	2026-03-22 13:47:55.712264083 +0800
+++ b/sdl2/src/video/mmiyoo/SDL_event_mmiyoo.c	2026-03-22 13:48:35.752264448 +0800
@@ -97,8 +97,8 @@
     #define L1      15
     #define R1      14
     #define START   28
-    #define SELECT  97
-    #define MENU    1
+    #define SELECT  1
+    #define MENU    97
 #endif
 
 #ifdef PANDORA
diff -uprN a/sdl2/src/video/mmiyoo/SDL_video_mmiyoo.c b/sdl2/src/video/mmiyoo/SDL_video_mmiyoo.c
--- a/sdl2/src/video/mmiyoo/SDL_video_mmiyoo.c	2026-07-06 11:13:53.102051570 +0800
+++ b/sdl2/src/video/mmiyoo/SDL_video_mmiyoo.c	2026-07-06 11:16:39.722051322 +0800
@@ -3613,21 +3613,21 @@ int GFX_Copy(const void *pixels, SDL_Rec
         uint16_t *src = (uint16_t *)pixels;
 
         asm volatile (
-            "1:  add %1, #64            ;"
+            "1:  add %0, #16            ;"
             "    vldmia %0!, {q0-q7}    ;"
-            "    vldmia %0!, {q8-q15}   ;"
+            "    vldmia %0!, {q8-q14}   ;"
             "    vstmia %1!, {q0-q7}    ;"
-            "    vstmia %1!, {q8-q15}   ;"
+            "    vstmia %1!, {q8-q14}   ;"
             "    vldmia %0!, {q0-q7}    ;"
-            "    vldmia %0!, {q8-q15}   ;"
+            "    vldmia %0!, {q8-q14}   ;"
             "    vstmia %1!, {q0-q7}    ;"
-            "    vstmia %1!, {q8-q15}   ;"
-            "    add %1, #64            ;"
+            "    vstmia %1!, {q8-q14}   ;"
+            "    add %0, #16            ;"
             "    subs %2, #1            ;"
             "    bne 1b                 ;"
             :
             : "r"(src), "r"(dst), "r"(192)
-            : "r8", "q0", "q1", "q2", "q3", "q4", "q5", "q6", "q7", "q8", "q9", "q10", "q11", "q12", "q13", "q14", "q15", "memory", "cc"
+            : "q0", "q1", "q2", "q3", "q4", "q5", "q6", "q7", "q8", "q9", "q10", "q11", "q12", "q13", "q14", "memory", "cc"
         );
     }
     else if ((srcrect.w == FB_W) && (srcrect.h == FB_H)) {
diff -uprN a/sdl2/src/video/mmiyoo/SDL_video_mmiyoo.h b/sdl2/src/video/mmiyoo/SDL_video_mmiyoo.h
--- a/sdl2/src/video/mmiyoo/SDL_video_mmiyoo.h	2026-03-22 12:56:43.522236042 +0800
+++ b/sdl2/src/video/mmiyoo/SDL_video_mmiyoo.h	2026-03-22 12:57:29.052236457 +0800
@@ -142,7 +142,7 @@
 #endif
 
 #ifdef FUNKEYS
-    #define DEF_FB_W                320
+    #define DEF_FB_W                240
     #define DEF_FB_H                240
     #define FB_BPP                  2
     #define IMG_W                   640
