diff -uprN a/build_opk.sh b/build_opk.sh
--- a/build_opk.sh	1970-01-01 08:00:00.000000000 +0800
+++ b/build_opk.sh	2026-03-08 19:10:12.624051059 +0800
@@ -0,0 +1,6 @@
+#!/bin/bash
+
+make -f Makefile.bittboy
+/opt/mmiyoo/bin/arm-linux-gnueabihf-strip yatka -o _opk/yatka
+mv music sfx skins _opk/
+mksquashfs _opk/* ../_opk/yatka_funkey-s.opk -all-root -no-xattrs -noappend -no-exports
diff -uprN a/Makefile.bittboy b/Makefile.bittboy
--- a/Makefile.bittboy	2026-02-16 13:08:31.896652598 +0800
+++ b/Makefile.bittboy	2026-02-16 13:21:45.626656717 +0800
@@ -1,6 +1,6 @@
 .PHONY: all clean zip
 
-BIN_BASE = /opt/miyoo/bin/
+BIN_BASE = /opt/mmiyoo/bin/
 PROJECT = yatka
 ZIPNAME = $(PROJECT).zip
 SRC = src/main.c src/data_persistence.c src/video.c src/sound.c \
@@ -8,9 +8,9 @@ SRC = src/main.c src/data_persistence.c
 	src/state_mainmenu.c src/skin.c
 OBJ = $(SRC:.c=.o)
 DEP = $(SRC:.c=.d)
-CFLAGS = -Iinc -D_BITTBOY -Ofast -march=armv5te -mtune=arm926ej-s
+CFLAGS = -Iinc -D_BITTBOY -Ofast -march=armv7-a -mtune=cortex-a7
 LDFLAGS = -s $(shell $(BIN_BASE)pkg-config --libs sdl SDL_image SDL_ttf SDL_mixer) -ljpeg -logg -lfreetype -lpng -lz -lbz2
-CC = arm-linux-gcc
+CC = /opt/mmiyoo/bin/arm-linux-gnueabihf-gcc
 
 all: $(PROJECT)
 
diff -uprN a/inc/main.h b/inc/main.h
--- a/inc/main.h	2026-02-16 13:09:41.446652959 +0800
+++ b/inc/main.h	2026-02-16 13:14:45.236654536 +0800
@@ -17,15 +17,15 @@
 
 // default keys
 #if defined _BITTBOY
-#define KEY_LEFT			SDLK_LEFT
-#define KEY_RIGHT			SDLK_RIGHT
-#define KEY_SOFTDROP		SDLK_DOWN
-#define KEY_HARDDROP		SDLK_LALT
-#define KEY_ROTATE_CW		SDLK_LCTRL
-#define KEY_ROTATE_CCW		SDLK_SPACE
-#define KEY_HOLD			SDLK_LSHIFT
-#define KEY_PAUSE			SDLK_RETURN
-#define KEY_QUIT			SDLK_ESCAPE
+#define KEY_LEFT			SDLK_LEFT 
+#define KEY_RIGHT			SDLK_RIGHT 
+#define KEY_SOFTDROP		SDLK_DOWN 
+#define KEY_HARDDROP		SDLK_LCTRL
+#define KEY_ROTATE_CW		SDLK_SPACE
+#define KEY_ROTATE_CCW		SDLK_LSHIFT
+#define KEY_HOLD			SDLK_LALT
+#define KEY_PAUSE			SDLK_RETURN 
+#define KEY_QUIT			SDLK_ESCAPE 
 #elif defined _RETROFW
 #define KEY_LEFT			SDLK_LEFT
 #define KEY_RIGHT			SDLK_RIGHT
diff -uprN a/src/state_mainmenu.c b/src/state_mainmenu.c
--- a/src/state_mainmenu.c	2026-02-16 13:10:47.876653304 +0800
+++ b/src/state_mainmenu.c	2026-02-16 13:16:23.306655045 +0800
@@ -89,7 +89,7 @@ static SDLKey getKey(void)
 			exit(0);
 		}
 	}
-	return SDLK_ESCAPE;
+	return SDLK_RCTRL;
 }
 
 static void text(int x, int y, const char *string, int alignx, int aligny)
@@ -418,8 +418,8 @@ static void action(void)
 		do
 		{
 			answer = getKey();
-		} while (SDLK_UP != answer && SDLK_DOWN != answer);
-		if (SDLK_UP == answer)
+		} while (SDLK_UP!= answer && SDLK_DOWN!= answer);
+		if (SDLK_UP== answer)
 		{
 			kleft = keys[KI_LEFT];
 			kright = keys[KI_RIGHT];
@@ -491,16 +491,16 @@ void mainmenu_processInputEvents(void)
 			case SDL_KEYDOWN:
 				switch (event.key.keysym.sym)
 				{
-					case SDLK_UP:
+					case SDLK_UP: 
 						up();
 						break;
-					case SDLK_RIGHT:
+					case SDLK_RIGHT: 
 						right();
 						break;
-					case SDLK_DOWN:
+					case SDLK_DOWN: 
 						down();
 						break;
-					case SDLK_LEFT:
+					case SDLK_LEFT: 
 						left();
 						break;
 					default:
diff -uprN a/src/state_settings.c b/src/state_settings.c
--- a/src/state_settings.c	2026-02-16 13:10:24.406653182 +0800
+++ b/src/state_settings.c	2026-02-16 13:16:50.516655186 +0800
@@ -281,19 +281,19 @@ void settings_processInputEvents(void)
 			case SDL_KEYDOWN:
 				switch (event.key.keysym.sym)
 				{
-					case SDLK_UP:
+					case SDLK_UP: 
 					{
 						up();
 					} break;
-					case SDLK_DOWN:
+					case SDLK_DOWN: 
 					{
 						down();
 					} break;
-					case SDLK_LEFT:
+					case SDLK_LEFT: 
 					{
 						left();
 					} break;
-					case SDLK_RIGHT:
+					case SDLK_RIGHT: 
 					{
 						right();
 					} break;
