diff -uprN a/build_opk.sh b/build_opk.sh
--- a/build_opk.sh	1970-01-01 08:00:00.000000000 +0800
+++ b/build_opk.sh	2026-04-25 11:43:40.947577826 +0800
@@ -0,0 +1,14 @@
+#!/bin/bash
+
+cp -r _opk/CMake/ _opk/Source/ ./
+rm -rf _opk/CMake/ _opk/Source/
+
+CMAKE_ARGS="-DTARGET_PLATFORM=miyoo_mini -DCMAKE_TOOLCHAIN_FILE=/opt/mmiyoo/share/buildroot/toolchainfile.cmake ${CMAKE_ARGS}"
+cmake $CMAKE_ARGS .
+make -j8 V=s
+cp -r devilutionx assets/ _opk/
+wget -P _opk/ https://github.com/diasurgical/devilutionx-assets/releases/download/v5/fonts.mpq
+#wget -P _opk/ https://github.com/diasurgical/devilutionx-assets/releases/download/v5/spawn.mpq
+wget -P _opk/ http://cdn.pvpgn.pro/diablo1/DIABDAT.MPQ
+#wget -P _opk/ http://cdn.pvpgn.pro/diablo1/hellfire/HELLFIRE.MPQ
+mksquashfs ./_opk/* ../_opk/devilutionx_funkey-s.opk -all-root -no-xattrs -noappend -no-exports
