diff -uprN a/build.sh b/build.sh
--- a/build.sh	1970-01-01 08:00:00.000000000 +0800
+++ b/build.sh	2026-03-25 10:49:42.061574001 +0800
@@ -0,0 +1,13 @@
+#!/bin/bash
+
+make -C phoneme_advanced_mr2/cdc/build/linux-arm-miyoo
+rm _opk/microbackend.jar
+rm -rf _opk/jvm/*
+rm -rf _opk/dist/*
+sed -i '148i #include "tbl.cc.h"' phoneme_advanced_mr2/cdc/build/linux-arm-miyoo/jcs/tbl.cc
+make -C phoneme_advanced_mr2/cdc/build/linux-arm-miyoo
+cd midpath && chmod +x ./*.sh
+./build.sh
+cd ..
+mv midpath/dist/* _opk/dist
+mv phoneme_advanced_mr2/cdc/build/linux-arm-miyoo/bin phoneme_advanced_mr2/cdc/build/linux-arm-miyoo/lib _opk/jvm
