diff -uprN a/Makefile.miyoo b/Makefile.miyoo
--- a/Makefile.miyoo	2026-02-18 10:21:47.192612037 +0800
+++ b/Makefile.miyoo	2026-02-18 10:27:24.052610209 +0800
@@ -1,12 +1,12 @@
 PRGNAME     = Zelda-Picross
 
 
-TOOLCHAIN = /opt/miyoo/usr
+TOOLCHAIN = /opt/mmiyoo
 EXESUFFIX = .dge
 
-CC = $(TOOLCHAIN)/bin/arm-linux-gcc
-CCP = $(TOOLCHAIN)/bin/arm-linux-g++
-LD = $(TOOLCHAIN)/bin/arm-linux-g++
+CC = $(TOOLCHAIN)/bin/arm-linux-gnueabihf-gcc
+CCP = $(TOOLCHAIN)/bin/arm-linux-gnueabihf-g++
+LD = $(TOOLCHAIN)/bin/arm-linux-gnueabihf-g++
 
 # add SDL dependencies
 SDL_LIB     = $(TOOLCHAIN)/lib
diff -uprN a/build_opk.sh b/build_opk.sh
--- a/build_opk.sh	1970-01-01 08:00:00.000000000 +0800
+++ b/build_opk.sh	2026-03-09 09:43:35.761973797 +0800
@@ -0,0 +1,6 @@
+#!/bin/bash
+mv _opk/Keyboard.cpp src/
+make -f Makefile.miyoo
+/opt/mmiyoo/bin/arm-linux-gnueabihf-strip Zelda-Picross.dge -o _opk/picross
+mv data _opk/
+mksquashfs _opk/* ../_opk/zeldapicross_funkey-s.opk -all-root -no-xattrs -noappend -no-exports
