diff -uprN a/build_opk.sh b/build_opk.sh
--- a/build_opk.sh	1970-01-01 08:00:00.000000000 +0800
+++ b/build_opk.sh	2026-03-07 20:41:55.102233619 +0800
@@ -0,0 +1,20 @@
+#!/bin/bash
+
+export PATH=$PATH:/opt/mmiyoo/bin
+export TARGETMACH=arm-linux
+export BUILDMACH=i686-pc-linux-gnu
+export CROSS=arm-linux-gnueabihf
+export CC=${CROSS}-gcc
+export LD=${CROSS}-ld
+export AS=${CROSS}-as
+export CXX=${CROSS}-g++
+
+cd src
+make clean
+make
+cp -f just4qix ..
+cd ..
+/opt/mmiyoo/bin/arm-linux-gnueabihf-strip just4qix -o _opk/just4qix
+mv gfx/ sound/ _opk
+mksquashfs _opk/* ../_opk/hcl_funkey-s.opk -all-root -no-xattrs -noappend -no-exports
+cd ..
