diff -uprN a/def.h b/def.h
--- a/def.h	2026-03-08 13:18:40.721264878 +0800
+++ b/def.h	2026-03-08 13:20:35.511264255 +0800
@@ -64,31 +64,31 @@
 #define MYKEY_LEFT SDLK_LEFT
 #endif
 #ifndef MYKEY_LINEUP
-#define MYKEY_LINEUP SDLK_r
+#define MYKEY_LINEUP SDLK_LSHIFT
 #endif
 #ifndef MYKEY_LINERIGHT
-#define MYKEY_LINERIGHT SDLK_g
+#define MYKEY_LINERIGHT SDLK_SPACE
 #endif
 #ifndef MYKEY_LINEDOWN
-#define MYKEY_LINEDOWN SDLK_f
+#define MYKEY_LINEDOWN SDLK_LCTRL
 #endif 
 #ifndef MYKEY_LINELEFT
-#define MYKEY_LINELEFT SDLK_d
+#define MYKEY_LINELEFT SDLK_LALT
 #endif
 #ifndef MYKEY_QUIT
-#define MYKEY_QUIT SDLK_q
+#define MYKEY_QUIT SDLK_ESCAPE
 #endif
 #ifndef MYKEY_START
-#define MYKEY_START SDLK_s
+#define MYKEY_START SDLK_RETURN
 #endif
 #ifndef MYKEY_MOD
-#define MYKEY_MOD SDLK_l
+#define MYKEY_MOD SDLK_TAB
 #endif
 #ifndef MYKEY_UNDO
-#define MYKEY_UNDO SDLK_u
+#define MYKEY_UNDO SDLK_i
 #endif
 #ifndef MYKEY_SYSTEM
-#define MYKEY_SYSTEM SDLK_q
+#define MYKEY_SYSTEM SDLK_r
 #endif
 #ifndef MYKEY_PAGEUP
 #define MYKEY_PAGEUP SDLK_PAGEUP
@@ -97,16 +97,16 @@
 #define MYKEY_PAGEDOWN SDLK_PAGEDOWN
 #endif
 #ifndef MYKEY_OPEN
-#define MYKEY_OPEN SDLK_RETURN
+#define MYKEY_OPEN SDLK_o
 #endif
 #ifndef MYKEY_PARENT
 #define MYKEY_PARENT SDLK_BACKSPACE
 #endif
 #ifndef MYKEY_OPERATION
-#define MYKEY_OPERATION SDLK_a
+#define MYKEY_OPERATION SDLK_z
 #endif
 #ifndef MYKEY_SELECT
-#define MYKEY_SELECT SDLK_INSERT
+#define MYKEY_SELECT SDLK_RCTRL
 #endif
 #ifndef MYKEY_TRANSFER
 #define MYKEY_TRANSFER SDLK_w
diff -uprN a/build_opk.sh b/build_opk.sh
--- a/build_opk.sh	1970-01-01 08:00:00.000000000 +0800
+++ b/build_opk.sh	2026-03-08 12:55:11.491272523 +0800
@@ -0,0 +1,9 @@
+#!/bin/bash
+
+CMAKE=1
+CMAKE_ARGS="-DCMAKE_TOOLCHAIN_FILE=/opt/mmiyoo/share/buildroot/toolchainfile.cmake ${CMAKE_ARGS}"
+cmake $CMAKE_ARGS .
+make
+/opt/mmiyoo/bin/arm-linux-gnueabihf-strip slitherlink -o _opk/slitherlink
+mv res _opk/
+mksquashfs _opk/* ../_opk/slitherlink_funkey-s.opk -all-root -no-xattrs -noappend -no-exports
