diff -uprN a/build_opk.sh b/build_opk.sh
--- a/build_opk.sh	1970-01-01 08:00:00.000000000 +0800
+++ b/build_opk.sh	2026-03-07 13:44:58.426329607 +0800
@@ -0,0 +1,6 @@
+#!/bin/bash
+
+make -f Makefile.gcw0
+/opt/mmiyoo/bin/arm-linux-gnueabihf-strip crafti -o ./_opk/crafti
+mksquashfs ./_opk/* ../_opk/crafti_funkey-s.opk -all-root -no-xattrs -noappend -no-exports
+
diff -uprN c/Makefile.gcw0 d/Makefile.gcw0
--- c/Makefile.gcw0	2025-02-04 13:10:34.312672440 +0800
+++ d/Makefile.gcw0	2025-02-04 13:08:46.002672601 +0800
@@ -1,11 +1,11 @@
-CXX	= /opt/gcw0-toolchain/bin/mipsel-linux-g++
-CXXFLAGS = -Ofast -mips32r2 -fno-rtti -fgcse-sm -fgcse-las -ffunction-sections -fdata-sections -DNDEBUG -Isrc -std=gnu++17  -Ingl -I. -Ipc
+CXX	= /opt/mmiyoo/bin/arm-linux-gnueabihf-g++
+CXXFLAGS = -Ofast -march=armv7-a+neon-vfpv4 -mtune=cortex-a7 -mfpu=neon-vfpv4 -fno-rtti -fgcse-sm -fgcse-las -ffunction-sections -fdata-sections -DNDEBUG -Isrc -std=gnu++17  -Ingl -I. -Ipc
 
 #Including OPUS Libraries
-LDFLAGS = -lc -lm -lstdc++ -lSDL -lSDL_mixer -Wl,--gc-sections -flto
+LDFLAGS = -lSDL_mixer -logg -lvorbisidec -lc -lm -lstdc++ -lSDL -lasound -Wl,--as-needed -Wl,--gc-sections -flto -s
 
 DEFINES = -Wall
-OUTPUT = crafti.elf
+OUTPUT = crafti
 
 OBJS += $(patsubst %.cpp, %.o, $(shell find . -name \*.cpp))
 
diff -uprN c/main.cpp d/main.cpp
--- c/main.cpp	2025-02-04 13:10:34.312672440 +0800
+++ d/main.cpp	2025-02-04 13:08:46.002672601 +0800
@@ -95,6 +95,8 @@ int main(int argc, char *argv[])
     terrainUninit();
     
     Close_Sound();
+
+    SDL_Quit();
 
     return 0;
 }
diff -uprN a/pc/gl.h b/pc/gl.h
--- a/pc/gl.h	2026-03-23 10:33:07.736002267 +0800
+++ b/pc/gl.h	2026-03-23 10:33:25.366002428 +0800
@@ -17,7 +17,7 @@
 #define SCREEN_WIDTH 240
 #define SCREEN_HEIGHT 240
 #else
-#define SCREEN_WIDTH 320
+#define SCREEN_WIDTH 240
 #define SCREEN_HEIGHT 240
 #endif
 
diff -uprN a/pc/libndls.h b/pc/libndls.h
--- a/pc/libndls.h	2026-03-08 09:06:45.781346882 +0800
+++ b/pc/libndls.h	2026-03-08 09:16:42.371343645 +0800
@@ -96,22 +96,22 @@
     #define KEY_NSPIRE_1 SDLK_TAB
 #endif   
   
-	#define KEY_NSPIRE_PERIOD SDLK_LSHIFT
+	#define KEY_NSPIRE_PERIOD SDLK_RETURN
 
 #ifdef NOJOY
     #define KEY_NSPIRE_UP SDLK_BACKSPACE
     #define KEY_NSPIRE_DOWN SDLK_TAB
 #else
-    #define KEY_NSPIRE_UP 555
-    #define KEY_NSPIRE_DOWN 556
+    #define KEY_NSPIRE_UP SDLK_UP
+    #define KEY_NSPIRE_DOWN SDLK_DOWN
 #endif
 
     #define KEY_NSPIRE_LEFT SDLK_LEFT
     #define KEY_NSPIRE_RIGHT SDLK_RIGHT
      
-    #define KEY_NSPIRE_DOC SDLK_SPACE
-    #define KEY_NSPIRE_5 SDLK_SPACE
-    #define KEY_NSPIRE_CLICK SDLK_LCTRL
+    #define KEY_NSPIRE_DOC SDLK_RCTRL
+    #define KEY_NSPIRE_5 SDLK_RCTRL
+    #define KEY_NSPIRE_CLICK SDLK_SPACE
     
 #ifdef NOJOY
 	#ifdef RG99
@@ -123,13 +123,13 @@
     #define KEY_NSPIRE_ESC SDLK_ESCAPE
 #endif
 
-    #define KEY_NSPIRE_MENU SDLK_RETURN
+    #define KEY_NSPIRE_MENU SDLK_HOME
 
-    #define KEY_NSPIRE_7 SDLK_LCTRL
-    #define KEY_NSPIRE_9 SDLK_LALT
+    #define KEY_NSPIRE_7 SDLK_SPACE
+    #define KEY_NSPIRE_9 SDLK_LCTRL
     
-    #define KEY_NSPIRE_MINUS 0
-    #define KEY_NSPIRE_PLUS 0
+    #define KEY_NSPIRE_MINUS SDLK_LSHIFT
+    #define KEY_NSPIRE_PLUS SDLK_LALT
     #define KEY_NSPIRE_0 0
     
     #define KEY_NSPIRE_CTRL 0
diff -uprN a/pc/task.cpp b/pc/task.cpp
--- a/pc/task.cpp	2026-03-08 09:19:49.221342631 +0800
+++ b/pc/task.cpp	2026-03-08 09:22:34.961341732 +0800
@@ -41,7 +41,7 @@ bool Task::keyPressed(int key)
 	SDL_PollEvent(&event);
 	
 	#ifndef NOJOY
-	if (key == 555 || key == 556 || key == SDLK_LEFT || key==SDLK_RIGHT)
+	if (key == 555 || key == 556)
 	{
 		if (joy_cord[0].x < -8000 && key == SDLK_LEFT)
 		{
