diff -uprN "a/Chinese (CN)" "b/Chinese (CN)"
--- "a/Chinese (CN)"	1970-01-01 08:00:00.000000000 +0800
+++ "b/Chinese (CN)"	2026-02-28 16:52:32.873016353 +0800
@@ -0,0 +1,116 @@
+Launch an application=启动应用程序
+Configure GMenu2X's options=配置GMenu2X的选项
+Skin=界面主题
+Configure skin=配置界面主题
+Wallpaper=系统壁纸
+Change GMenu2X wallpaper=更换系统壁纸
+Log Viewer=日志阅读器
+Displays last launched program's output=显示最后运行的程序输出日志
+About=关于
+Info about GMenu2X=关于GMenu2X
+Do you want to delete the log file?=您是否想删除此日志文件?
+Yes=是
+No=否
+Select an application=选择想要运行的程序
+Settings=设置
+Language=系统语言
+Set the language used by GMenu2X=设置系统界面语言
+Save last selection=记录最后选择的分组
+Save the last selected link and section on exit=记录最后运行的程序和分组
+Output logs=输出日志
+Logs the output of the links. Use the Log Viewer to read them.=记录并输出日志
+Trim extensions=拓展名设置
+Trim extensions in the file explorer.=文件管理器拓展名设置
+Screen Timeout=屏幕超时
+Set screen's backlight timeout in seconds=设置屏幕背光超时时间（秒）
+Button repeat rate=按钮重复率
+Set button repetitions per second=设置按钮每秒重复次数
+Set the skin used by GMenu2X=为系统选择一个界面主题
+Top Bar=顶部栏颜色
+Color of the top bar=设置顶部栏颜色
+Bottom Bar=底部栏颜色
+Color of the bottom bar=设置底部栏颜色
+Selection=选中项颜色
+Color of the selection and other interface details=当前被选中项目的颜色
+Message Box=对话框颜色
+Background color of the message box=设置对话框的背景颜色
+Message Box Border=对话框边框
+Border color of the message box=设置对话框边框的颜色
+Message Box Selection=对话框中当前被选中的项目
+Color of the selection of the message box=设置对话框中当前被选中项目的颜色
+Title=程序标题
+Link title=设置程序名称
+Description=程序概述
+Link description=设置程序的简要概述
+Section=程序分组
+The section this link belongs to=选择程序所属的分组
+Icon=程序图标
+Select an icon for this link=选择程序的图标
+Manual=程序手册
+Select a manual or README file=选择程序使用手册或说明文档
+Selector Directory=选择程序目录
+Directory to scan for the selector=选择程序默认加载目录(如:游戏rom存放目录)
+Selector Browser=程序浏览器
+Allow the selector to change directory=允许程序选择器改变目录
+Clock frequency=时钟频率
+CPU clock frequency for this link=此程序的CPU时钟频率
+Selector Filter=选择程序筛选器
+Selector filter (Separate values with a comma)=选择程序筛选(用逗号分割)
+Display Console=显示终端
+Must be enabled for console-based applications=基于控制台的应用程序必须启用
+Are you sure?=是否确定?
+Insert a name for the new section=为新分组输入名称
+Up one folder=返回上层目录
+Select=选择
+Confirm=确认
+Exit=退出
+Add section=新建分组
+Delete section=删除分组
+Add link in $1=添加链接 $1
+Show manual of $1=显示$1说明
+Edit $1=编辑 $1
+Delete $1=删除 $1
+Deleting $1=正在删除 $1
+Cancel=取消
+OK=确定
+Backspace=删除
+Space=空格
+Change keys=修改按键
+Change page=换页
+Page=页
+Launching $1=正在运行 $1
+Switch=切换
+Clear=清除
+Decrease=减小
+Increase=增大
+Change value=改变值
+Change color component=改变颜色组合
+Edit=编辑
+no items=目录为空
+Scroll=滚动
+Untitled=无标题
+Wallpaper selection=壁纸选择窗口
+Select a wallpaper from the list=从下列选择一个图片作为壁纸
+Lng=ch
+applications=应用
+settings=设置
+emulators=模拟器
+games=游戏
+Explorer=文件浏览器
+VOLUME=音量
+BRIGHTNESS=背光
+POWERDOWN=关机
+SET LAUNCHER=前端切换
+EJECT=断开
+MOUNT=挂载
+In progress...=操作中...
+Saving...=保存中...
+Loading...=加载中...
+Are you sure ?=你确定?
+Shutting down...=关机...
+CONTROLS=操作
+A: Launch link / Confirm action=A: 启动应用/确认操作
+B: Show this help menu=B: 显示此帮助菜单
+L, R: Change section=L, R: 切换界面
+SELECT: Show contextual menu=SELECT: 显示上下文菜单
+START: Show options menu=START: 显示选项菜单
diff -uprN a/build.sh b/build.sh
--- a/build.sh	1970-01-01 08:00:00.000000000 +0800
+++ b/build.sh	2026-02-27 09:14:15.760803649 +0800
@@ -0,0 +1,6 @@
+#!/bin/bash
+
+CMAKE=1
+CMAKE_ARGS="-DCMAKE_TOOLCHAIN_FILE=/opt/FunKey-sdk-2.3.0/share/buildroot/toolchainfile.cmake ${CMAKE_ARGS}"
+cmake $CMAKE_ARGS .
+make
diff -uprN a/buildopts.h.cmakein b/buildopts.h.cmakein
--- a/buildopts.h.cmakein	2026-02-27 09:21:45.970804751 +0800
+++ b/buildopts.h.cmakein	2026-02-27 09:15:12.560803788 +0800
@@ -1,7 +1,7 @@
 #ifndef GMENU2X_BUILDOPTS_H
 #define GMENU2X_BUILDOPTS_H
 
-#define GMENU2X_SYSTEM_DIR "@CMAKE_INSTALL_FULL_DATADIR@/gmenu2x"
+#define GMENU2X_SYSTEM_DIR "/usr/share/gmenu2x"
 #define GMENU2X_CARD_ROOT "@CARD_ROOT@"
 
 #endif /* GMENU2X_BUILDOPTS_H */
diff -uprN a/src/contextmenu.cpp b/src/contextmenu.cpp
--- a/src/contextmenu.cpp	2026-02-28 18:52:03.185473568 +0800
+++ b/src/contextmenu.cpp	2026-02-27 09:16:36.350803993 +0800
@@ -82,7 +82,7 @@ ContextMenu::ContextMenu(GMenu2X &gmenu2
 	for (auto option : options) {
 		w = std::max(w, font.getTextWidth(option->text));
 	}
-	w += 23;
+	w += 80;
 	const int h = (font.getLineSpacing() + 2) * options.size() + 8;
 	box = {
 		static_cast<Sint16>((gmenu2x.width() - w) / 2),
diff -uprN a/src/funkeymenu.cpp b/src/funkeymenu.cpp
--- a/src/funkeymenu.cpp	2026-02-28 18:52:03.185473568 +0800
+++ b/src/funkeymenu.cpp	2026-02-28 18:53:24.895474636 +0800
@@ -1,4 +1,5 @@
 #include "funkeymenu.h"
+#include "gmenu2x.h"
 #include <iostream>
 
 /// -------------- DEFINES --------------
@@ -32,15 +33,15 @@
 #define MENU_BG_SQURE_WIDTH         180
 #define MENU_BG_SQUREE_HEIGHT       140
 
-#define MENU_FONT_NAME_TITLE        "/usr/games/menu_resources/OpenSans-Bold.ttf"
+#define MENU_FONT_NAME_TITLE        "/usr/share/fonts/truetype/dejavu/DejaVuSans-Bold.ttf"
 #define MENU_FONT_SIZE_TITLE        22
-#define MENU_FONT_NAME_INFO         "/usr/games/menu_resources/OpenSans-Bold.ttf"
+#define MENU_FONT_NAME_INFO         "/usr/share/fonts/truetype/dejavu/DejaVuSans-Bold.ttf"
 #define MENU_FONT_SIZE_INFO         16
-#define MENU_FONT_NAME_SMALL_INFO   "/usr/games/menu_resources/OpenSans-Regular.ttf"
+#define MENU_FONT_NAME_SMALL_INFO   "/usr/share/fonts/truetype/dejavu/DejaVuSans-Bold.ttf"
 #define MENU_FONT_SIZE_SMALL_INFO   13
-#define MENU_PNG_BG_PATH            "/usr/games/menu_resources/zone_bg.png"
-#define MENU_PNG_ARROW_TOP_PATH     "/usr/games/menu_resources/arrow_top.png"
-#define MENU_PNG_ARROW_BOTTOM_PATH  "/usr/games/menu_resources/arrow_bottom.png"
+#define MENU_PNG_BG_PATH            "/usr/share/gmenu2x/skins/320x240/Default/imgs/zone_bg.png"
+#define MENU_PNG_ARROW_TOP_PATH     "/usr/share/gmenu2x/skins/320x240/Default/imgs/arrow_top.png"
+#define MENU_PNG_ARROW_BOTTOM_PATH  "/usr/share/gmenu2x/skins/320x240/Default/imgs/arrow_bottom.png"
 
 #define GRAY_MAIN_R                 85
 #define GRAY_MAIN_G                 85
@@ -279,13 +280,17 @@ void FunkeyMenu::add_menu_zone(ENUM_MENU
 	SDL_Surface *text_surface = NULL;
 	SDL_Surface *surface = menu_zone_surfaces[nb_menu_zones-1];
 	SDL_Rect text_pos;
+	std::string volume_str = gmenu2x_ptr->tr["VOLUME"];
+	std::string bright_str = gmenu2x_ptr->tr["BRIGHTNESS"];
+	std::string power_str = gmenu2x_ptr->tr["POWERDOWN"];
+	std::string launch_str = gmenu2x_ptr->tr["SET LAUNCHER"];
 
 	/// --------- Add new zone ---------
 	switch(menu_type){
 	case MENU_TYPE_VOLUME:
 		MENU_DEBUG_PRINTF("Init MENU_TYPE_VOLUME\n");
 		/// ------ Text ------
-		text_surface = TTF_RenderText_Blended(menu_title_font, "VOLUME", text_color);
+		text_surface = TTF_RenderUTF8_Solid(menu_title_font, volume_str.c_str(), text_color);
 		text_pos.x = (surface->w - MENU_ZONE_WIDTH)/2 + (MENU_ZONE_WIDTH - text_surface->w)/2;
 		text_pos.y = surface->h - MENU_ZONE_HEIGHT/2 - text_surface->h/2 - padding_y_from_center_menu_zone;
 		SDL_BlitSurface(text_surface, NULL, surface, &text_pos);
@@ -298,7 +303,7 @@ void FunkeyMenu::add_menu_zone(ENUM_MENU
 	case MENU_TYPE_BRIGHTNESS:
 		MENU_DEBUG_PRINTF("Init MENU_TYPE_BRIGHTNESS\n");
 		/// ------ Text ------
-		text_surface = TTF_RenderText_Blended(menu_title_font, "BRIGHTNESS", text_color);
+		text_surface = TTF_RenderUTF8_Solid(menu_title_font, bright_str.c_str(), text_color);
 		text_pos.x = (surface->w - MENU_ZONE_WIDTH)/2 + (MENU_ZONE_WIDTH - text_surface->w)/2;
 		text_pos.y = surface->h - MENU_ZONE_HEIGHT/2 - text_surface->h/2 - padding_y_from_center_menu_zone;
 		SDL_BlitSurface(text_surface, NULL, surface, &text_pos);
@@ -311,7 +316,7 @@ void FunkeyMenu::add_menu_zone(ENUM_MENU
 	case MENU_TYPE_SAVE:
 		MENU_DEBUG_PRINTF("Init MENU_TYPE_SAVE\n");
 		/// ------ Text ------
-		text_surface = TTF_RenderText_Blended(menu_title_font, "SAVE", text_color);
+		text_surface = TTF_RenderUTF8_Solid(menu_title_font, "SAVE", text_color);
 		text_pos.x = (surface->w - MENU_ZONE_WIDTH)/2 + (MENU_ZONE_WIDTH - text_surface->w)/2;
 		text_pos.y = surface->h - MENU_ZONE_HEIGHT/2 - text_surface->h/2 - padding_y_from_center_menu_zone*2;
 		SDL_BlitSurface(text_surface, NULL, surface, &text_pos);
@@ -319,7 +324,7 @@ void FunkeyMenu::add_menu_zone(ENUM_MENU
 	case MENU_TYPE_LOAD:
 		MENU_DEBUG_PRINTF("Init MENU_TYPE_LOAD\n");
 		/// ------ Text ------
-		text_surface = TTF_RenderText_Blended(menu_title_font, "LOAD", text_color);
+		text_surface = TTF_RenderUTF8_Solid(menu_title_font, "LOAD", text_color);
 		text_pos.x = (surface->w - MENU_ZONE_WIDTH)/2 + (MENU_ZONE_WIDTH - text_surface->w)/2;
 		text_pos.y = surface->h - MENU_ZONE_HEIGHT/2 - text_surface->h/2 - padding_y_from_center_menu_zone*2;
 		SDL_BlitSurface(text_surface, NULL, surface, &text_pos);
@@ -327,7 +332,7 @@ void FunkeyMenu::add_menu_zone(ENUM_MENU
 	case MENU_TYPE_ASPECT_RATIO:
 		MENU_DEBUG_PRINTF("Init MENU_TYPE_ASPECT_RATIO\n");
 		/// ------ Text ------
-		text_surface = TTF_RenderText_Blended(menu_title_font, "ASPECT RATIO", text_color);
+		text_surface = TTF_RenderUTF8_Solid(menu_title_font, "ASPECT RATIO", text_color);
 		text_pos.x = (surface->w - MENU_ZONE_WIDTH)/2 + (MENU_ZONE_WIDTH - text_surface->w)/2;
 		text_pos.y = surface->h - MENU_ZONE_HEIGHT/2 - text_surface->h/2 - padding_y_from_center_menu_zone;
 		SDL_BlitSurface(text_surface, NULL, surface, &text_pos);
@@ -335,7 +340,7 @@ void FunkeyMenu::add_menu_zone(ENUM_MENU
 	case MENU_TYPE_EXIT:
 		MENU_DEBUG_PRINTF("Init MENU_TYPE_EXIT\n");
 		/// ------ Text ------
-		text_surface = TTF_RenderText_Blended(menu_title_font, "EXIT GAME", text_color);
+		text_surface = TTF_RenderUTF8_Solid(menu_title_font, "EXIT GAME", text_color);
 		text_pos.x = (surface->w - MENU_ZONE_WIDTH)/2 + (MENU_ZONE_WIDTH - text_surface->w)/2;
 		text_pos.y = surface->h - MENU_ZONE_HEIGHT/2 - text_surface->h/2;
 		SDL_BlitSurface(text_surface, NULL, surface, &text_pos);
@@ -343,7 +348,7 @@ void FunkeyMenu::add_menu_zone(ENUM_MENU
 	case MENU_TYPE_USB:
 		MENU_DEBUG_PRINTF("Init MENU_TYPE_USB\n");
 		/// ------ Text ------
-		/*text_surface = TTF_RenderText_Blended(menu_title_font, "USB", text_color);
+		/*text_surface = TTF_RenderUTF8_Solid(menu_title_font, "USB", text_color);
 		text_pos.x = (surface->w - MENU_ZONE_WIDTH)/2 + (MENU_ZONE_WIDTH - text_surface->w)/2;
 		text_pos.y = surface->h - MENU_ZONE_HEIGHT/2 - text_surface->h/2;
 		SDL_BlitSurface(text_surface, NULL, surface, &text_pos);*/
@@ -351,12 +356,12 @@ void FunkeyMenu::add_menu_zone(ENUM_MENU
 	case MENU_TYPE_LAUNCHER:
 		MENU_DEBUG_PRINTF("Init MENU_TYPE_LAUNCHER\n");
 		/// ------ Text ------
-		text_surface = TTF_RenderText_Blended(menu_title_font, "SET LAUNCHER", text_color);
+		text_surface = TTF_RenderUTF8_Solid(menu_title_font, launch_str.c_str(), text_color);
 		text_pos.x = (surface->w - MENU_ZONE_WIDTH)/2 + (MENU_ZONE_WIDTH - text_surface->w)/2;
 		text_pos.y = surface->h - MENU_ZONE_HEIGHT/2 - text_surface->h/2 - padding_y_from_center_menu_zone*2;
 		SDL_BlitSurface(text_surface, NULL, surface, &text_pos);
 		/// ------ Text ------
-		text_surface = TTF_RenderText_Blended(menu_title_font, "RETROFE", text_color);
+		text_surface = TTF_RenderUTF8_Solid(menu_title_font, "RETROFE", text_color);
 		text_pos.x = (surface->w - MENU_ZONE_WIDTH)/2 + (MENU_ZONE_WIDTH - text_surface->w)/2;
 		text_pos.y = surface->h - MENU_ZONE_HEIGHT/2 - text_surface->h/2;
 		SDL_BlitSurface(text_surface, NULL, surface, &text_pos);
@@ -364,7 +369,7 @@ void FunkeyMenu::add_menu_zone(ENUM_MENU
 	case MENU_TYPE_POWERDOWN:
 		MENU_DEBUG_PRINTF("Init MENU_TYPE_POWERDOWN\n");
 		/// ------ Text ------
-		text_surface = TTF_RenderText_Blended(menu_title_font, "POWERDOWN", text_color);
+		text_surface = TTF_RenderUTF8_Solid(menu_title_font, power_str.c_str(), text_color);
 		text_pos.x = (surface->w - MENU_ZONE_WIDTH)/2 + (MENU_ZONE_WIDTH - text_surface->w)/2;
 		text_pos.y = surface->h - MENU_ZONE_HEIGHT/2 - text_surface->h/2;
 		SDL_BlitSurface(text_surface, NULL, surface, &text_pos);
@@ -394,7 +399,7 @@ void FunkeyMenu::init_menu_zones(){
 	/// Init USB Menu
 	add_menu_zone(MENU_TYPE_USB);
 	/// Init Launcher Menu
-	add_menu_zone(MENU_TYPE_LAUNCHER);
+	//add_menu_zone(MENU_TYPE_LAUNCHER);
 	/// Init Powerdown Menu
 	add_menu_zone(MENU_TYPE_POWERDOWN);
 }
@@ -486,6 +491,13 @@ void FunkeyMenu::menu_screen_refresh(int
 	SDL_Rect menu_blit_window;
 	menu_blit_window.x = 0;
 	menu_blit_window.w = SCREEN_HORIZONTAL_SIZE;
+	std::string mount_str = gmenu2x_ptr->tr["MOUNT"];
+	std::string eject_str = gmenu2x_ptr->tr["EJECT"];
+	std::string progress_str = gmenu2x_ptr->tr["In progress..."];
+	std::string save_str = gmenu2x_ptr->tr["Saving..."];
+	std::string load_str = gmenu2x_ptr->tr["Loading..."];
+	std::string sure_str = gmenu2x_ptr->tr["Are you sure ?"];
+	std::string shut_str = gmenu2x_ptr->tr["Shutting down..."];
 
 	/// --------- Blit prev menu Zone going away ----------
 	menu_blit_window.y = scroll;
@@ -534,19 +546,19 @@ void FunkeyMenu::menu_screen_refresh(int
 		case MENU_TYPE_SAVE:
 			/// ---- Write slot -----
 			sprintf(text_tmp, "IN SLOT   < %d >", savestate_slot+1);
-			text_surface = TTF_RenderText_Blended(menu_info_font, text_tmp, text_color);
+			text_surface = TTF_RenderUTF8_Solid(menu_info_font, text_tmp, text_color);
 			text_pos.x = (draw_screen->w - MENU_ZONE_WIDTH)/2 + (MENU_ZONE_WIDTH - text_surface->w)/2;
 			text_pos.y = draw_screen->h - MENU_ZONE_HEIGHT/2 - text_surface->h/2;
 			SDL_BlitSurface(text_surface, NULL, draw_screen, &text_pos);
 
 			if(menu_action){
-				sprintf(text_tmp, "Saving...");
-				text_surface = TTF_RenderText_Blended(menu_info_font, text_tmp, text_color);
+				sprintf(text_tmp, save_str.c_str());
+				text_surface = TTF_RenderUTF8_Solid(menu_info_font, text_tmp, text_color);
 			}
 			else{
 				if(menu_confirmation){
-					sprintf(text_tmp, "Are you sure ?");
-					text_surface = TTF_RenderText_Blended(menu_info_font, text_tmp, text_color);
+					sprintf(text_tmp, sure_str.c_str());
+					text_surface = TTF_RenderUTF8_Solid(menu_info_font, text_tmp, text_color);
 				}
 				else{
 					/// ---- Write current Save state ----
@@ -560,19 +572,19 @@ void FunkeyMenu::menu_screen_refresh(int
 		case MENU_TYPE_LOAD:
 			/// ---- Write slot -----
 			sprintf(text_tmp, "FROM SLOT   < %d >", savestate_slot+1);
-			text_surface = TTF_RenderText_Blended(menu_info_font, text_tmp, text_color);
+			text_surface = TTF_RenderUTF8_Solid(menu_info_font, text_tmp, text_color);
 			text_pos.x = (draw_screen->w - MENU_ZONE_WIDTH)/2 + (MENU_ZONE_WIDTH - text_surface->w)/2;
 			text_pos.y = draw_screen->h - MENU_ZONE_HEIGHT/2 - text_surface->h/2;
 			SDL_BlitSurface(text_surface, NULL, draw_screen, &text_pos);
 
 			if(menu_action){
-				sprintf(text_tmp, "Loading...");
-				text_surface = TTF_RenderText_Blended(menu_info_font, text_tmp, text_color);
+				sprintf(text_tmp, load_str.c_str());
+				text_surface = TTF_RenderUTF8_Solid(menu_info_font, text_tmp, text_color);
 			}
 			else{
 				if(menu_confirmation){
-					sprintf(text_tmp, "Are you sure ?");
-					text_surface = TTF_RenderText_Blended(menu_info_font, text_tmp, text_color);
+					sprintf(text_tmp, sure_str.c_str());
+					text_surface = TTF_RenderUTF8_Solid(menu_info_font, text_tmp, text_color);
 				}
 				else{
 					/// ---- Write current Load state ----
@@ -585,7 +597,7 @@ void FunkeyMenu::menu_screen_refresh(int
 
 		case MENU_TYPE_ASPECT_RATIO:
 			sprintf(text_tmp, "<   %s   >", aspect_ratio_name[aspect_ratio]);
-			text_surface = TTF_RenderText_Blended(menu_info_font, text_tmp, text_color);
+			text_surface = TTF_RenderUTF8_Solid(menu_info_font, text_tmp, text_color);
 			text_pos.x = (draw_screen->w - MENU_ZONE_WIDTH)/2 + (MENU_ZONE_WIDTH - text_surface->w)/2;
 			text_pos.y = draw_screen->h - MENU_ZONE_HEIGHT/2 - text_surface->h/2 + padding_y_from_center_menu_zone;
 			SDL_BlitSurface(text_surface, NULL, draw_screen, &text_pos);
@@ -593,22 +605,22 @@ void FunkeyMenu::menu_screen_refresh(int
 
 		case MENU_TYPE_USB:
 			/// ---- Write slot -----
-			sprintf(text_tmp, "%s USB", usb_sharing?"EJECT":"MOUNT");
-			text_surface = TTF_RenderText_Blended(menu_title_font, text_tmp, text_color);
+			sprintf(text_tmp, "%s USB", usb_sharing?eject_str.c_str():mount_str.c_str());
+			text_surface = TTF_RenderUTF8_Solid(menu_title_font, text_tmp, text_color);
 			text_pos.x = (draw_screen->w - MENU_ZONE_WIDTH)/2 + (MENU_ZONE_WIDTH - text_surface->w)/2;
 			text_pos.y = draw_screen->h - MENU_ZONE_HEIGHT/2 - text_surface->h/2;
 			SDL_BlitSurface(text_surface, NULL, draw_screen, &text_pos);
 
 			if(menu_action){
-				sprintf(text_tmp, "in progress ...");
-				text_surface = TTF_RenderText_Blended(menu_info_font, text_tmp, text_color);
+				sprintf(text_tmp, progress_str.c_str());
+				text_surface = TTF_RenderUTF8_Solid(menu_info_font, text_tmp, text_color);
 				text_pos.x = (draw_screen->w - MENU_ZONE_WIDTH)/2 + (MENU_ZONE_WIDTH - text_surface->w)/2;
 				text_pos.y = draw_screen->h - MENU_ZONE_HEIGHT/2 - text_surface->h/2 + 2*padding_y_from_center_menu_zone;
 				SDL_BlitSurface(text_surface, NULL, draw_screen, &text_pos);
 			}
 			else if(menu_confirmation){
-				sprintf(text_tmp, "Are you sure ?");
-				text_surface = TTF_RenderText_Blended(menu_info_font, text_tmp, text_color);
+				sprintf(text_tmp, sure_str.c_str());
+				text_surface = TTF_RenderUTF8_Solid(menu_info_font, text_tmp, text_color);
 				text_pos.x = (draw_screen->w - MENU_ZONE_WIDTH)/2 + (MENU_ZONE_WIDTH - text_surface->w)/2;
 				text_pos.y = draw_screen->h - MENU_ZONE_HEIGHT/2 - text_surface->h/2 + 2*padding_y_from_center_menu_zone;
 				SDL_BlitSurface(text_surface, NULL, draw_screen, &text_pos);
@@ -620,15 +632,15 @@ void FunkeyMenu::menu_screen_refresh(int
 
 		case MENU_TYPE_LAUNCHER:
 			if(menu_action){
-				sprintf(text_tmp, "In progress...");
-				text_surface = TTF_RenderText_Blended(menu_info_font, text_tmp, text_color);
+				sprintf(text_tmp, progress_str.c_str());
+				text_surface = TTF_RenderUTF8_Solid(menu_info_font, text_tmp, text_color);
 				text_pos.x = (draw_screen->w - MENU_ZONE_WIDTH)/2 + (MENU_ZONE_WIDTH - text_surface->w)/2;
 				text_pos.y = draw_screen->h - MENU_ZONE_HEIGHT/2 - text_surface->h/2 + 2*padding_y_from_center_menu_zone;
 				SDL_BlitSurface(text_surface, NULL, draw_screen, &text_pos);
 			}
 			else if(menu_confirmation){
-				sprintf(text_tmp, "Are you sure ?");
-				text_surface = TTF_RenderText_Blended(menu_info_font, text_tmp, text_color);
+				sprintf(text_tmp, sure_str.c_str());
+				text_surface = TTF_RenderUTF8_Solid(menu_info_font, text_tmp, text_color);
 				text_pos.x = (draw_screen->w - MENU_ZONE_WIDTH)/2 + (MENU_ZONE_WIDTH - text_surface->w)/2;
 				text_pos.y = draw_screen->h - MENU_ZONE_HEIGHT/2 - text_surface->h/2 + 2*padding_y_from_center_menu_zone;
 				SDL_BlitSurface(text_surface, NULL, draw_screen, &text_pos);
@@ -638,16 +650,16 @@ void FunkeyMenu::menu_screen_refresh(int
 		case MENU_TYPE_EXIT:
 		case MENU_TYPE_POWERDOWN:
 			if(menu_action){
-				sprintf(text_tmp, "Shutting down...");
-				text_surface = TTF_RenderText_Blended(menu_info_font, text_tmp, text_color);
+				sprintf(text_tmp, shut_str.c_str());
+				text_surface = TTF_RenderUTF8_Solid(menu_info_font, text_tmp, text_color);
 				text_pos.x = (draw_screen->w - MENU_ZONE_WIDTH)/2 + (MENU_ZONE_WIDTH - text_surface->w)/2;
 				text_pos.y = draw_screen->h - MENU_ZONE_HEIGHT/2 - text_surface->h/2 + 2*padding_y_from_center_menu_zone;
 				SDL_BlitSurface(text_surface, NULL, draw_screen, &text_pos);
 			}
 			else{
 				if(menu_confirmation){
-					sprintf(text_tmp, "Are you sure ?");
-					text_surface = TTF_RenderText_Blended(menu_info_font, text_tmp, text_color);
+					sprintf(text_tmp, sure_str.c_str());
+					text_surface = TTF_RenderUTF8_Solid(menu_info_font, text_tmp, text_color);
 					text_pos.x = (draw_screen->w - MENU_ZONE_WIDTH)/2 + (MENU_ZONE_WIDTH - text_surface->w)/2;
 					text_pos.y = draw_screen->h - MENU_ZONE_HEIGHT/2 - text_surface->h/2 + 2*padding_y_from_center_menu_zone;
 					SDL_BlitSurface(text_surface, NULL, draw_screen, &text_pos);
@@ -684,8 +696,8 @@ void FunkeyMenu::menu_screen_refresh(int
 	if(SDL_BlitSurface(draw_screen, NULL, hw_screen, NULL)){
 		MENU_ERROR_PRINTF("ERROR Could not bit draw_screen to hw_screen: %s\n", SDL_GetError());
 	}
-	SDL_Flip(hw_screen);
-	//gmenu2x_ptr->s->flip();
+//	SDL_Flip(hw_screen);
+	gmenu2x_ptr->s->flip();
 }
 
 
diff -uprN a/src/gmenu2x.cpp b/src/gmenu2x.cpp
--- a/src/gmenu2x.cpp	2026-02-28 18:52:03.185473568 +0800
+++ b/src/gmenu2x.cpp	2026-03-01 10:39:38.436337253 +0800
@@ -76,7 +76,7 @@
 #define DEFAULT_FONT_SIZE 12
 
 #ifndef DEFAULT_FALLBACK_FONTS
-#define DEFAULT_FALLBACK_FONTS ,{"/usr/share/fonts/truetype/droid/DroidSansFallbackFull.ttf",13},{"/usr/share/fonts/truetype/droid/DroidSansFallback.ttf",13}
+#define DEFAULT_FALLBACK_FONTS ,{"/usr/share/fonts/truetype/droid/DroidSansFallbackFull.ttf",12},{"/usr/share/fonts/truetype/droid/DroidSansFallback.ttf",12}
 #endif
 
 using namespace std;
@@ -427,7 +427,7 @@ void GMenu2X::initMenu() {
 
 	// Add action links in the applications section.
 	auto appIdx = menu->sectionNamed("applications");
-	menu->addActionLink(appIdx, "Explorer",
+	menu->addActionLink(appIdx, tr["Explorer"],
 			bind(&GMenu2X::explorer, this),
 			tr["Launch an application"],
 			"skin:icons/explorer.png");
diff -uprN a/src/menu.cpp b/src/menu.cpp
--- a/src/menu.cpp	2026-02-28 18:52:03.175473568 +0800
+++ b/src/menu.cpp	2026-02-28 13:13:56.917118000 +0800
@@ -171,7 +171,7 @@ void Menu::updateSectionTextSurfaces() {
 	const auto &font = *gmenu2x.font;
 	section_text_surfaces.resize(sections.size());
 	for (std::size_t i = 0; i < sections.size(); ++i)
-		section_text_surfaces[i] = font.render(sections[i]);
+		section_text_surfaces[i] = font.render(gmenu2x.tr[sections[i]]);
 }
 
 void Menu::calcSectionRange(int &leftSection, int &rightSection) {
diff -uprN a/src/selector.cpp b/src/selector.cpp
--- a/src/selector.cpp	2026-02-28 18:52:03.185473568 +0800
+++ b/src/selector.cpp	2026-02-28 20:04:16.545530185 +0800
@@ -117,7 +117,7 @@ int Selector::exec(int startSelection) {
 				string path = screendir + trimExtension(fl[selected]) + ".png";
 				auto screenshot = OffscreenSurface::loadImage(path, false);
 				if (screenshot) {
-					screenshot->blitRight(s, gmenu2x.width(), 0, gmenu2x.width(), gmenu2x.height(), 128u);
+					screenshot->blitRight(s, gmenu2x.width(), 25, 170, 190, 255u);
 				}
 			}
 
diff -uprN a/src/surface.cpp b/src/surface.cpp
--- a/src/surface.cpp	2026-03-01 10:32:03.346334891 +0800
+++ b/src/surface.cpp	2026-03-01 10:54:55.266342011 +0800
@@ -29,6 +29,16 @@
 #include <cassert>
 #include <iomanip>
 #include <utility>
+#include <sys/ioctl.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <linux/fb.h>
+
+void wait_for_HW() {
+	int fd0 = open("/dev/fb0", O_RDWR);
+	ioctl(fd0, FBIO_WAITFORVSYNC, 0);
+	close(fd0);
+}
 
 using namespace std;
 
@@ -356,5 +366,6 @@ unique_ptr<OutputSurface> OutputSurface:
 }
 
 void OutputSurface::flip() {
+    wait_for_HW();
 	SDL_Flip(raw);
 }
