diff -uprN a/build.sh b/build.sh
--- a/build.sh	1970-01-01 08:00:00.000000000 +0800
+++ b/build.sh	2025-06-30 10:15:52.661204397 +0800
@@ -0,0 +1,7 @@
+#!/bin/bash
+
+export ARCH=arm
+export CROSS_COMPILE=/opt/FunKey-sdk-2.3.0/bin/arm-funkey-linux-gnueabihf-
+
+make -j1 V=s
+#make modules_install INSTALL_MOD_PATH="bin"
\ No newline at end of file
diff -uprN a/drivers/mtd/spi/spi-nor-ids.c b/drivers/mtd/spi/spi-nor-ids.c
--- a/drivers/mtd/spi/spi-nor-ids.c	2025-06-30 10:17:30.665184707 +0800
+++ b/drivers/mtd/spi/spi-nor-ids.c	2025-06-30 10:17:50.968351559 +0800
@@ -89,6 +89,11 @@ const struct flash_info spi_nor_ids[] =
 #ifdef CONFIG_SPI_FLASH_GIGADEVICE	/* GIGADEVICE */
 	/* GigaDevice */
 	{
+		INFO("gd25q08", 0xc84014, 0, 64 * 1024,  16,
+			SECT_4K | SPI_NOR_DUAL_READ | SPI_NOR_QUAD_READ |
+			SPI_NOR_HAS_LOCK | SPI_NOR_HAS_TB)
+	},
+	{
 		INFO("gd25q16", 0xc84015, 0, 64 * 1024,  32,
 			SECT_4K | SPI_NOR_DUAL_READ | SPI_NOR_QUAD_READ |
 			SPI_NOR_HAS_LOCK | SPI_NOR_HAS_TB)
