diff -uprN a/cannonball/Makefile.funkey b/cannonball/Makefile.funkey
--- a/cannonball/Makefile.funkey	2025-01-21 15:40:30.000000000 +0800
+++ b/cannonball/Makefile.funkey	2025-01-21 15:40:13.000000000 +0800
@@ -1,9 +1,9 @@
-CHAINPREFIX=/opt/FunKey-sdk
-CROSS_COMPILE=$(CHAINPREFIX)/usr/bin/arm-funkey-linux-musleabihf-
+CHAINPREFIX=/opt/FunKey-sdk-2.3.0
+CROSS_COMPILE=$(CHAINPREFIX)/usr/bin/arm-funkey-linux-gnueabihf-
 CC = $(CROSS_COMPILE)gcc
 CXX = $(CROSS_COMPILE)g++
 STRIP = $(CROSS_COMPILE)strip
-SYSROOT     := $(CHAINPREFIX)/usr/arm-funkey-linux-musleabihf/sysroot
+SYSROOT     := $(CHAINPREFIX)/usr/arm-funkey-linux-gnueabihf/sysroot
 SDL_CFLAGS  := $(shell $(SYSROOT)/usr/bin/sdl-config --cflags)
 SDL_LIBS    := $(shell $(SYSROOT)/usr/bin/sdl-config --libs)
 
@@ -45,4 +45,4 @@ $(OUTPUT):	${OBJS}
 	
 clean:
 	rm *.o engine/audio/*.o engine/*.o hwvideo/*.o frontend/*.o hwaudio/*.o sdl/*.o thirdparty/crc/*.o thirdparty/sxmlc/*.o cannonboard/*.o font/*.o
-	rm ${OUTPUT}
\ No newline at end of file
+	rm ${OUTPUT}
