diff -uprN a/build.sh b/build.sh
--- a/build.sh	1970-01-01 08:00:00.000000000 +0800
+++ b/build.sh	2025-02-07 16:10:36.368518549 +0800
@@ -0,0 +1,3 @@
+#!/bin/sh
+
+make -C openbor-6391/engine BUILD_OPENDINGUX=1
diff -uprN a/openbor-6391/engine/Makefile b/openbor-6391/engine/Makefile
--- a/openbor-6391/engine/Makefile	2025-02-07 15:57:26.997656790 +0800
+++ b/openbor-6391/engine/Makefile	2025-02-07 15:56:18.412188499 +0800
@@ -15,13 +15,33 @@
 #----------------------------------------------------------------------------------------------------
 
 ifndef VERSION_NAME
-VERSION_NAME = OpenBOR
+VERSION_NAME = openbor
 endif
 
 #----------------------------------------------------------------------------------------------------
 # Defines
 #----------------------------------------------------------------------------------------------------
 
+ifdef BUILD_OPENDINGUX
+TARGET 	        = $(VERSION_NAME).elf
+TARGET_FINAL    = $(VERSION_NAME)
+TARGET_PLATFORM = OPENDINGUX
+BUILD_SDL       = 1
+BUILD_GFX       = 1
+BUILD_PTHREAD   = 1
+BUILD_SDL_IO    = 1
+BUILD_OPENGL    = 1
+BUILD_LOADGL    = 1
+BUILD_VORBIS    = 1
+BUILDING        = 1
+CC              = /opt/mmiyoo/bin/arm-linux-gnueabihf-gcc
+INCLUDES        = /opt/mmiyoo/arm-buildroot-linux-gnueabihf/sysroot/usr/include/ \
+                  /opt/mmiyoo/arm-buildroot-linux-gnueabihf/sysroot/usr/include/SDL2 ./source/webmlib/
+LIBRARIES       = /opt/mmiyoo/arm-buildroot-linux-gnueabihf/sysroot/usr/lib/
+ifeq ($(BUILD_OPENDINGUX), 0)
+BUILD_DEBUG     = 1
+endif
+endif
 
 ifdef BUILD_PSP
 TARGET          = $(VERSION_NAME)
@@ -201,6 +221,9 @@ endif
 ifdef BUILD_PANDORA
 STRIP 	        = $(PNDDEV)/bin/arm-none-linux-gnueabi-strip $(TARGET) -o $(TARGET_FINAL)
 endif
+ifdef BUILD_OPENDINGUX
+STRIP           = /opt/mmiyoo/bin/arm-linux-gnueabihf-strip $(TARGET) -o $(TARGET_FINAL)
+endif
 ifdef BUILD_WII
 STRIP           = elf2dol $< $@
 endif
@@ -443,7 +466,7 @@ OBJS            = $(MAIN)
 #----------------------------------------------------------------------------------------------------
 
 CFLAGS 	       += $(addprefix -I", $(addsuffix ", $(INCS))) $(ARCHFLAGS) -D$(TARGET_PLATFORM)
-CFLAGS 	       += -g -Wall -Werror -fsigned-char -std=gnu99
+CFLAGS 	       += -g -Wall -fsigned-char -std=gnu99
 
 
 ifndef BUILD_DARWIN
@@ -485,6 +508,10 @@ ifdef BUILD_PANDORA
 CFLAGS         += -DLINUX
 endif
 
+ifdef BUILD_OPENDINGUX
+INCS           += '/opt/mmiyoo/arm-buildroot-linux-gnueabihf/sysroot/usr/bin/libpng-config --prefix'/include/libpng16
+endif
+
 
 ifdef BUILD_WII
 CFLAGS 	       += -DCACHE_BACKGROUNDS -DREVERSE_COLOR -D__ppc__ $(MACHDEP) -Wl,-Map,$(TARGET_MAP),-wrap,wiiuse_register
@@ -551,6 +578,11 @@ LIBS 	       += -lpspgu -lpspaudio -lpsp
 endif
 
 
+ifdef BUILD_OPENDINGUX
+CFLAGS += -O2 -fomit-frame-pointer -ffast-math -funroll-loops -Wall -Wno-unknown-pragmas -Wno-format -DOPENDINGUX -marm -mtune=cortex-a7 -mfpu=neon-vfpv4 -mfloat-abi=hard -march=armv7-a
+endif
+
+
 ifdef BUILD_DARWIN
 LIBS           += -Wl,-syslibroot,$(SDKPATH) \
                   -framework Cocoa \
diff -uprN a/openbor-6391/engine/sdl/joysticks.h b/openbor-6391/engine/sdl/joysticks.h
--- a/openbor-6391/engine/sdl/joysticks.h	2025-02-07 15:57:50.340795122 +0800
+++ b/openbor-6391/engine/sdl/joysticks.h	2025-02-07 15:55:42.133527664 +0800
@@ -22,18 +22,18 @@
 #define JOY_NAME_SIZE      1 + 1 + JOY_MAX_INPUTS * JOY_LIST_TOTAL
 
 #ifdef OPENDINGUX
-#define OPENDINGUX_BUTTON_UP     SDLK_UP
-#define OPENDINGUX_BUTTON_DOWN   SDLK_DOWN
-#define OPENDINGUX_BUTTON_RIGHT  SDLK_RIGHT
-#define OPENDINGUX_BUTTON_LEFT   SDLK_LEFT
-#define OPENDINGUX_BUTTON_R      SDLK_BACKSPACE
-#define OPENDINGUX_BUTTON_L      SDLK_TAB
-#define OPENDINGUX_BUTTON_A      SDLK_LCTRL
-#define OPENDINGUX_BUTTON_B      SDLK_LALT
-#define OPENDINGUX_BUTTON_X      SDLK_SPACE
-#define OPENDINGUX_BUTTON_Y      SDLK_LSHIFT
-#define OPENDINGUX_BUTTON_SELECT SDLK_ESCAPE
-#define OPENDINGUX_BUTTON_START  SDLK_RETURN
+#define OPENDINGUX_BUTTON_UP     SDL_SCANCODE_UP
+#define OPENDINGUX_BUTTON_DOWN   SDL_SCANCODE_DOWN
+#define OPENDINGUX_BUTTON_RIGHT  SDL_SCANCODE_RIGHT
+#define OPENDINGUX_BUTTON_LEFT   SDL_SCANCODE_LEFT
+#define OPENDINGUX_BUTTON_L      SDL_SCANCODE_BACKSPACE
+#define OPENDINGUX_BUTTON_R      SDL_SCANCODE_TAB
+#define OPENDINGUX_BUTTON_A      SDL_SCANCODE_LCTRL
+#define OPENDINGUX_BUTTON_B      SDL_SCANCODE_LALT
+#define OPENDINGUX_BUTTON_X      SDL_SCANCODE_SPACE
+#define OPENDINGUX_BUTTON_Y      SDL_SCANCODE_LSHIFT
+#define OPENDINGUX_BUTTON_SELECT SDL_SCANCODE_ESCAPE
+#define OPENDINGUX_BUTTON_START  SDL_SCANCODE_RETURN
 #endif
 
 /* Real-Time Joystick Data */
diff -uprN a/openbor-6391/engine/sdl/sdlport.c b/openbor-6391/engine/sdl/sdlport.c
--- a/openbor-6391/engine/sdl/sdlport.c	2025-02-07 15:57:50.344794974 +0800
+++ b/openbor-6391/engine/sdl/sdlport.c	2025-02-07 15:58:35.183139847 +0800
@@ -27,10 +27,18 @@ char packfile[MAX_FILENAME_LEN] = {"bor.
 #include <unistd.h>
 char rootDir[MAX_BUFFER_LEN] = {""};
 #endif
+
+#ifdef OPENDINGUX
+char paksDir[MAX_FILENAME_LEN] = {"/mnt/OpenBOR/Paks"};
+char savesDir[MAX_FILENAME_LEN] = {"/mnt/OpenBOR/Saves"};
+char logsDir[MAX_FILENAME_LEN] = {"/mnt/OpenBOR/Logs"};
+char screenShotsDir[MAX_FILENAME_LEN] = {"/mnt/OpenBOR/ScreenShots"};
+#else
 char paksDir[MAX_FILENAME_LEN] = {"Paks"};
 char savesDir[MAX_FILENAME_LEN] = {"Saves"};
 char logsDir[MAX_FILENAME_LEN] = {"Logs"};
 char screenShotsDir[MAX_FILENAME_LEN] = {"ScreenShots"};
+#endif
 
 // sleeps for the given number of microseconds
 #if _POSIX_C_SOURCE >= 199309L
