diff -uprN c/build.sh d/build.sh
--- c/build.sh	1970-01-01 08:00:00.000000000 +0800
+++ d/build.sh	2025-02-06 08:53:40.008535090 +0800
@@ -0,0 +1,6 @@
+#!/usr/bin/env bash
+
+cp out/bin/ffplay _opk/
+cp out/lib/libavcodec.so.55 out/lib/libavdevice.so.55 out/lib/libavfilter.so.3 out/lib/libavformat.so.55 out/lib/libavutil.so.52 out/lib/libpostproc.so.52 out/lib/libswresample.so.0 out/lib/libswscale.so.2 _opk/lib/
+rm ffplay-gcw0/BUILD_CONFIG.txt ffplay-gcw0/SOURCE.txt ffplay-gcw0/LICENSE.txt ffplay-gcw0/quvi ffplay-gcw0/ffplay-s*
+cp -r ffplay-gcw0/* _opk/
diff -uprN c/cfg.sh d/cfg.sh
--- c/cfg.sh	1970-01-01 08:00:00.000000000 +0800
+++ d/cfg.sh	2025-02-06 08:53:40.008535090 +0800
@@ -0,0 +1,24 @@
+#!/usr/bin/env bash
+
+./configure --prefix=$PWD/out \
+            --enable-cross-compile \
+            --cross-prefix="/opt/FunKey-sdk-2.3.0/bin/arm-funkey-linux-gnueabihf-" \
+            --pkg-config="/opt/FunKey-sdk-2.3.0/bin/pkg-config" \
+            --target-os=linux \
+            --arch=arm \
+            --extra-cflags="-Os -march=armv7-a+neon-vfpv4 -mtune=cortex-a7 -mfpu=neon-vfpv4 -I/opt/FunKey-sdk-2.3.0/arm-funkey-linux-gnueabihf/sysroot/usr/bin/../../usr/include/SDL -D_GNU_SOURCE=1 -D_REENTRANT -D_FILE_OFFSET_BITS=64 -Wno-unused-result -fPIC" \
+            --extra-ldflags="-L/opt/FunKey-sdk-2.3.0/arm-funkey-linux-gnueabihf/sysroot/usr/bin/../../usr/lib -lSDL -lpthread -lpng  -lSDL_ttf -lSDL_image -ldl -lm -lpthread -lz" \
+            --enable-shared \
+            --enable-small \
+            --disable-static \
+            --disable-network \
+            --disable-libcdio \
+            --disable-armv5te \
+            --disable-armv6 \
+            --disable-armv6t2 \
+            --disable-stripping \
+            --disable-debug \
+            --disable-ffmpeg \
+            --disable-ffserver \
+            --disable-doc \
+            --enable-gpl
diff -uprN c/ffplay-gcw0/ffplay.dge d/ffplay-gcw0/ffplay.dge
--- c/ffplay-gcw0/ffplay.dge	2025-02-06 08:56:31.408536655 +0800
+++ d/ffplay-gcw0/ffplay.dge	2025-02-06 08:55:14.818535956 +0800
@@ -119,11 +119,11 @@ fi
 
 if [ -d "$HOME" ]
 then
-        mkdir -p "$HOME"/.config/ffplay/file_settings
+        mkdir -p "$HOME"/.ffplay/file_settings
 
-        if [ ! -d "$HOME/video/Qbertaddict1's YouTube Channel" ]
+        if [ ! -d "$HOME/.ffplay/video/Qbertaddict1's YouTube Channel" ]
         then
-                cp -r "video" "$HOME"
+                cp -r "video" "$HOME/.ffplay"
         fi
 fi
 
diff -uprN c/ffplay-gcw0/ffplay.funkey-s.desktop d/ffplay-gcw0/ffplay.funkey-s.desktop
--- c/ffplay-gcw0/ffplay.funkey-s.desktop	1970-01-01 08:00:00.000000000 +0800
+++ d/ffplay-gcw0/ffplay.funkey-s.desktop	2025-02-06 08:55:14.818535956 +0800
@@ -0,0 +1,16 @@
+[Desktop Entry]
+Name=FFplay
+Name[ch]=FFplay
+Version=2.1.4
+Comment=Simple media player
+Comment[ch]=轻量级媒体播放器
+Exec=ffplay.dge %f
+Terminal=false
+Type=Application
+StartupNotify=true
+Icon=ffplay
+Categories=applications;
+SelectorDir=/mnt/Media
+SelectorFilter=mp4,mp3,mpa,ogg,flac,rmvb,avi
+MimeType=video/avi;video/mp4;audio/mp4;video/ogg;audio/ogg;video/x-matroska;audio/x-matroska;audio/x-flac;audio/mpeg;video/mpeg;video/x-flv;video/x-ms-asf;video/quicktime;video/3gpp;video/webm;audio/vnd.rn-realaudio;audio/x-pn-realaudio;audio/x-pn-realaudio-plugin;video/vnd.rn-realvideo;application/vnd.rn-realmedia;application/vnd.rn-realmedia-vbr;audio/x-mpegurl;
+X-OD-Manual=README.txt
diff -uprN c/ffplay-gcw0/ffplay.gcw0.desktop d/ffplay-gcw0/ffplay.gcw0.desktop
--- c/ffplay-gcw0/ffplay.gcw0.desktop	2025-02-06 08:56:31.348536654 +0800
+++ d/ffplay-gcw0/ffplay.gcw0.desktop	1970-01-01 08:00:00.000000000 +0800
@@ -1,12 +0,0 @@
-[Desktop Entry]
-Name=FFplay
-Version=2.1.4
-Comment=Simple media player
-Exec=ffplay.dge %f
-Terminal=false
-Type=Application
-StartupNotify=true
-Icon=ffplay
-Categories=applications;
-MimeType=video/avi;video/mp4;audio/mp4;video/ogg;audio/ogg;video/x-matroska;audio/x-matroska;audio/x-flac;audio/mpeg;video/mpeg;video/x-flv;video/x-ms-asf;video/quicktime;video/3gpp;video/webm;audio/vnd.rn-realaudio;audio/x-pn-realaudio;audio/x-pn-realaudio-plugin;video/vnd.rn-realvideo;application/vnd.rn-realmedia;application/vnd.rn-realmedia-vbr;audio/x-mpegurl;
-X-OD-Manual=README.txt
diff -uprN c/ffplay.c d/ffplay.c
--- c/ffplay.c	2025-02-06 08:56:22.398536573 +0800
+++ d/ffplay.c	2025-02-06 08:54:38.968535629 +0800
@@ -3236,17 +3236,17 @@ static void event_loop(VideoState *cur_s
                 cur_stream->force_refresh = 1;
                 break;
             case SDLK_p:
-            case SDLK_SPACE:
+            case SDLK_x:
                 toggle_pause(cur_stream);
                 break;
             case SDLK_s: // S: Step to next frame
                 step_to_next_frame(cur_stream);
                 break;
             case SDLK_TAB: // GCW0: Left shoulder button.
-            case SDLK_a:
+            case SDLK_k:
                 stream_cycle_channel(cur_stream, AVMEDIA_TYPE_AUDIO);
                 break;
-            case SDLK_LCTRL: // GCW0: A button.
+            case SDLK_a: // GCW0: A button.
             case SDLK_v:
                 stream_cycle_channel(cur_stream, AVMEDIA_TYPE_VIDEO);
                 break;
@@ -3255,30 +3255,30 @@ static void event_loop(VideoState *cur_s
                 stream_cycle_channel(cur_stream, AVMEDIA_TYPE_AUDIO);
                 stream_cycle_channel(cur_stream, AVMEDIA_TYPE_SUBTITLE);
                 break;
-            case SDLK_BACKSPACE: // GCW0: Right shoulder button.
+            case SDLK_b: // GCW0: Right shoulder button.
             case SDLK_t:
                 stream_cycle_channel(cur_stream, AVMEDIA_TYPE_SUBTITLE);
                 break;
-            case SDLK_LSHIFT: // GCW0: X button.
+            case SDLK_y: // GCW0: X button.
             case SDLK_w:
                 toggle_audio_display(cur_stream);
                 break;
-            case SDLK_PAGEUP:
+            case SDLK_m:
                 incr = 600.0;
                 goto do_seek;
-            case SDLK_PAGEDOWN:
+            case SDLK_n:
                 incr = -600.0;
                 goto do_seek;
-            case SDLK_LEFT:
+            case SDLK_l:
                 incr = -10.0;
                 goto do_seek;
-            case SDLK_RIGHT:
+            case SDLK_r:
                 incr = 10.0;
                 goto do_seek;
-            case SDLK_UP:
+            case SDLK_u:
                 incr = 60.0;
                 goto do_seek;
-            case SDLK_DOWN:
+            case SDLK_d:
                 incr = -60.0;
             do_seek:
                     if (seek_by_bytes) {
