diff -uprN c/Makefile.gcw0 d/Makefile.gcw0
--- c/Makefile.gcw0	2025-02-04 13:10:34.312672440 +0800
+++ d/Makefile.gcw0	2025-02-04 13:08:46.002672601 +0800
@@ -1,11 +1,11 @@
-CXX	= /opt/gcw0-toolchain/bin/mipsel-linux-g++
-CXXFLAGS = -Ofast -mips32r2 -fno-rtti -fgcse-sm -fgcse-las -ffunction-sections -fdata-sections -DNDEBUG -Isrc -std=gnu++17  -Ingl -I. -Ipc
+CXX	= /opt/FunKey-sdk-2.3.0/bin/arm-funkey-linux-gnueabihf-g++
+CXXFLAGS = -Ofast -march=armv7-a+neon-vfpv4 -mtune=cortex-a7 -mfpu=neon-vfpv4 -fno-rtti -fgcse-sm -fgcse-las -ffunction-sections -fdata-sections -DNDEBUG -Isrc -std=gnu++17  -Ingl -I. -Ipc
 
 #Including OPUS Libraries
-LDFLAGS = -lc -lm -lstdc++ -lSDL -lSDL_mixer -Wl,--gc-sections -flto
+LDFLAGS = -lSDL_mixer -logg -lvorbisidec -lmikmod -lmpg123 -lc -lm -lstdc++ -lSDL -lasound -Wl,--as-needed -Wl,--gc-sections -flto -s
 
 DEFINES = -Wall
-OUTPUT = crafti.elf
+OUTPUT = crafti
 
 OBJS += $(patsubst %.cpp, %.o, $(shell find . -name \*.cpp))
 
diff -uprN c/main.cpp d/main.cpp
--- c/main.cpp	2025-02-04 13:10:34.312672440 +0800
+++ d/main.cpp	2025-02-04 13:08:46.002672601 +0800
@@ -95,6 +95,8 @@ int main(int argc, char *argv[])
     terrainUninit();
     
     Close_Sound();
+    
+    SDL_Quit();
 
     return 0;
 }
diff -uprN c/pc/libndls.h d/pc/libndls.h
--- c/pc/libndls.h	2025-02-04 13:35:25.912670219 +0800
+++ d/pc/libndls.h	2025-02-04 13:09:52.432672502 +0800
@@ -92,26 +92,26 @@
     #define KEY_NSPIRE_1 0
     #endif
 #else
-    #define KEY_NSPIRE_3 SDLK_BACKSPACE
-    #define KEY_NSPIRE_1 SDLK_TAB
+    #define KEY_NSPIRE_3 SDLK_n
+    #define KEY_NSPIRE_1 SDLK_m
 #endif   
   
-	#define KEY_NSPIRE_PERIOD SDLK_LSHIFT
+	#define KEY_NSPIRE_PERIOD SDLK_s
 
 #ifdef NOJOY
     #define KEY_NSPIRE_UP SDLK_BACKSPACE
     #define KEY_NSPIRE_DOWN SDLK_TAB
 #else
-    #define KEY_NSPIRE_UP 555
-    #define KEY_NSPIRE_DOWN 556
+    #define KEY_NSPIRE_UP SDLK_u
+    #define KEY_NSPIRE_DOWN SDLK_d
 #endif
 
-    #define KEY_NSPIRE_LEFT SDLK_LEFT
-    #define KEY_NSPIRE_RIGHT SDLK_RIGHT
+    #define KEY_NSPIRE_LEFT SDLK_l
+    #define KEY_NSPIRE_RIGHT SDLK_r
      
-    #define KEY_NSPIRE_DOC SDLK_SPACE
+    #define KEY_NSPIRE_DOC SDLK_k
     #define KEY_NSPIRE_5 SDLK_SPACE
-    #define KEY_NSPIRE_CLICK SDLK_LCTRL
+    #define KEY_NSPIRE_CLICK SDLK_a
     
 #ifdef NOJOY
 	#ifdef RG99
@@ -120,16 +120,16 @@
     #define KEY_NSPIRE_ESC SDLK_HOME
     #endif
 #else
-    #define KEY_NSPIRE_ESC SDLK_ESCAPE
+    #define KEY_NSPIRE_ESC SDLK_q
 #endif
 
-    #define KEY_NSPIRE_MENU SDLK_RETURN
+    #define KEY_NSPIRE_MENU SDLK_HOME
 
-    #define KEY_NSPIRE_7 SDLK_LCTRL
-    #define KEY_NSPIRE_9 SDLK_LALT
+    #define KEY_NSPIRE_7 SDLK_a
+    #define KEY_NSPIRE_9 SDLK_b
     
-    #define KEY_NSPIRE_MINUS 0
-    #define KEY_NSPIRE_PLUS 0
+    #define KEY_NSPIRE_MINUS SDLK_x
+    #define KEY_NSPIRE_PLUS SDLK_y
     #define KEY_NSPIRE_0 0
     
     #define KEY_NSPIRE_CTRL 0
@@ -139,4 +139,4 @@
 
 
     #endif
-#endif
\ No newline at end of file
+#endif
