diff -uprN a/Makefile b/Makefile
--- a/Makefile	2025-01-22 13:43:47.462287633 +0800
+++ b/Makefile	2025-01-22 13:40:38.977417754 +0800
@@ -1,8 +1,8 @@
 # Define the applications properties here:
 
-TARGET = ./dist/PocketSNES.dge
+TARGET = ./dist/pocketsnes
 
-CROSS_COMPILE ?= mipsel-linux-
+CROSS_COMPILE ?= /opt/FunKey-sdk-2.3.0/bin/arm-funkey-linux-gnueabihf-
 
 CC  := $(CROSS_COMPILE)gcc
 CXX := $(CROSS_COMPILE)g++
@@ -17,9 +17,9 @@ INCLUDE = -I pocketsnes \
 		-I pocketsnes/include \
 		-I menu -I pocketsnes/linux -I pocketsnes/snes9x
 
-CFLAGS = $(INCLUDE) -DRC_OPTIMIZED -DGCW_ZERO -DGCW_JOYSTICK -D__LINUX__ -D__DINGUX__ -DFOREVER_16_BIT -DLAGFIX
+CFLAGS = $(INCLUDE) -DRC_OPTIMIZED -DGCW_ZERO -DGCW_JOYSTICK -D__LINUX__ -DFOREVER_16_BIT -DLAGFIX
 # CFLAGS += -ggdb3 -Og
-CFLAGS += -Ofast -fdata-sections -ffunction-sections -mips32r2 -mno-mips16 -mplt -mno-shared
+CFLAGS += -Ofast -fdata-sections -ffunction-sections -march=armv7-a+neon-vfpv4 -mtune=cortex-a7 -mfpu=neon-vfpv4 
 CFLAGS += -fomit-frame-pointer -fno-builtin -fno-common -flto=4 -fno-unroll-loops
 CFLAGS += -DFAST_ALIGNED_LSB_WORD_ACCESS
 CFLAGS += $(SDL_CFLAGS)
diff -uprN a/sal/linux/sal.c b/sal/linux/sal.c
--- a/sal/linux/sal.c	2025-01-22 13:43:54.190033132 +0800
+++ b/sal/linux/sal.c	2025-01-22 13:43:00.516063545 +0800
@@ -129,19 +129,19 @@ static u32 sal_Input(int held)
 	inputHeld = 0;
 
 	u8 *keystate = SDL_GetKeyState(NULL);
-	if ( keystate[SDLK_LCTRL] )		inputHeld |= SAL_INPUT_A;
-	if ( keystate[SDLK_LALT] )		inputHeld |= SAL_INPUT_B;
-	if ( keystate[SDLK_SPACE] )		inputHeld |= SAL_INPUT_X;
-	if ( keystate[SDLK_LSHIFT] )	inputHeld |= SAL_INPUT_Y;
-	if ( keystate[SDLK_TAB] )		inputHeld |= SAL_INPUT_L;
-	if ( keystate[SDLK_BACKSPACE] )	inputHeld |= SAL_INPUT_R;
-	if ( keystate[SDLK_RETURN] )	inputHeld |= SAL_INPUT_START;
-	if ( keystate[SDLK_ESCAPE] )	inputHeld |= SAL_INPUT_SELECT;
-	if ( keystate[SDLK_UP] )		inputHeld |= SAL_INPUT_UP;
-	if ( keystate[SDLK_DOWN] )		inputHeld |= SAL_INPUT_DOWN;
-	if ( keystate[SDLK_LEFT] )		inputHeld |= SAL_INPUT_LEFT;
-	if ( keystate[SDLK_RIGHT] )		inputHeld |= SAL_INPUT_RIGHT;
-	if ( keystate[SDLK_HOME] )		inputHeld |= SAL_INPUT_MENU;
+	if ( keystate[SDLK_a] )		inputHeld |= SAL_INPUT_A;
+	if ( keystate[SDLK_b] )		inputHeld |= SAL_INPUT_B;
+	if ( keystate[SDLK_x] )		inputHeld |= SAL_INPUT_X;
+	if ( keystate[SDLK_y] )	inputHeld |= SAL_INPUT_Y;
+	if ( keystate[SDLK_m] )		inputHeld |= SAL_INPUT_L;
+	if ( keystate[SDLK_n] )	inputHeld |= SAL_INPUT_R;
+	if ( keystate[SDLK_s] )	inputHeld |= SAL_INPUT_START;
+	if ( keystate[SDLK_k] )	inputHeld |= SAL_INPUT_SELECT;
+	if ( keystate[SDLK_u] )		inputHeld |= SAL_INPUT_UP;
+	if ( keystate[SDLK_d] )		inputHeld |= SAL_INPUT_DOWN;
+	if ( keystate[SDLK_l] )		inputHeld |= SAL_INPUT_LEFT;
+	if ( keystate[SDLK_r] )		inputHeld |= SAL_INPUT_RIGHT;
+	if ( keystate[SDLK_q] )		inputHeld |= SAL_INPUT_MENU;
 
 #ifdef GCW_JOYSTICK
 	if (mMenuOptions->analogJoy) {
