diff -uprN c/Makefile.bittboy d/Makefile.bittboy
--- c/Makefile.bittboy	2025-02-04 11:17:05.482682576 +0800
+++ d/Makefile.bittboy	2025-02-04 11:15:11.722682746 +0800
@@ -1,5 +1,5 @@
 PRGNAME     = oswan
-CC			= /opt/bittboy-toolchain/bin/arm-linux-gcc
+CC			= /opt/FunKey-sdk-2.3.0/bin/arm-funkey-linux-gnueabihf-gcc
 
 PROFILE = 0
 
diff -uprN c/main/menu/menu.c d/main/menu/menu.c
--- c/main/menu/menu.c	2025-02-04 11:17:22.612682551 +0800
+++ d/main/menu/menu.c	2025-02-04 11:16:39.682682615 +0800
@@ -63,43 +63,43 @@ static const char* Return_Text_Button(ui
 	switch(button)
 	{
 		/* UP button */
-		case SDLK_UP:
+		case SDLK_u:
 			return "D-UP";
 		break;
 		/* DOWN button */
-		case SDLK_DOWN:
+		case SDLK_d:
 			return "D-DOWN";
 		break;
 		/* LEFT button */
-		case SDLK_LEFT:
+		case SDLK_l:
 			return "D-LEFT";
 		break;
 		/* RIGHT button */
-		case SDLK_RIGHT:
+		case SDLK_r:
 			return "D-RIGHT";
 		break;
 		/* A button */
-		case SDLK_LCTRL:
+		case SDLK_a:
 			return "A";
 		break;
 		/* B button */
-		case SDLK_LALT:
+		case SDLK_b:
 			return "B";
 		break;
 		/* X button */
-		case SDLK_LSHIFT:
+		case SDLK_x:
 			return "X";
 		break;
 		/* Y button */
-		case SDLK_SPACE:
+		case SDLK_y:
 			return "Y";
 		break;
 		/* L button */
-		case SDLK_TAB:
+		case SDLK_m:
 			return "L";
 		break;
 		/* R button */
-		case SDLK_BACKSPACE:
+		case SDLK_n:
 			return "R";
 		break;
 		case SDLK_PAGEUP:
@@ -108,10 +108,10 @@ static const char* Return_Text_Button(ui
 		case SDLK_PAGEDOWN:
 			return "R2";
 		break;
-		case SDLK_RETURN:
+		case SDLK_s:
 			return "Start";
 		break;
-		case SDLK_ESCAPE:
+		case SDLK_k:
 			return "Select";
 		break;
 		default:
@@ -308,15 +308,15 @@ static uint32_t sdl_controls_update_inpu
 		case 0:
 		switch(k)
 		{
-			case SDLK_UP:
+			case SDLK_u:
 				if (menu_oswan.Choose_Menu_value == 0) menu_oswan.Choose_Menu_value = menu_oswan.maximum_menu-1;
 				else menu_oswan.Choose_Menu_value--;
 			break;
-			case SDLK_DOWN:
+			case SDLK_d:
 				menu_oswan.Choose_Menu_value++;
 				if (menu_oswan.Choose_Menu_value > menu_oswan.maximum_menu-1) menu_oswan.Choose_Menu_value = 0;
 			break;
-			case SDLK_LEFT:
+			case SDLK_l:
 				if (menu_oswan.Choose_Menu_value == 1 || menu_oswan.Choose_Menu_value == 2)
 				{
 					if (menu_oswan.state_number == 0) menu_oswan.state_number = 9;
@@ -328,7 +328,7 @@ static uint32_t sdl_controls_update_inpu
 					else menu_oswan.scaling--;
 				}
 			break;
-			case SDLK_RIGHT:
+			case SDLK_r:
 				if (menu_oswan.Choose_Menu_value == 1 || menu_oswan.Choose_Menu_value == 2)
 				{
 					menu_oswan.state_number++;
@@ -340,8 +340,8 @@ static uint32_t sdl_controls_update_inpu
 					if (menu_oswan.scaling > 2) menu_oswan.scaling = 0;
 				}
 			break;
-			case SDLK_LCTRL:
-			case SDLK_RETURN:
+			case SDLK_a:
+			case SDLK_s:
 				if (menu_oswan.Choose_Menu_value == 4) Set_Menu(CONTROLS_MENU);
 				else if (menu_oswan.Choose_Menu_value == 5)
 				{
@@ -367,27 +367,27 @@ static uint32_t sdl_controls_update_inpu
 		case 1:
 		switch(k)
 		{	
-			case SDLK_UP:
+			case SDLK_u:
 				if (menu_oswan.Choose_Menu_value == 0) menu_oswan.Choose_Menu_value = menu_oswan.maximum_menu-1;
 				else menu_oswan.Choose_Menu_value--;
 			break;
-			case SDLK_DOWN:
+			case SDLK_d:
 				menu_oswan.Choose_Menu_value++;
 				if (menu_oswan.Choose_Menu_value > menu_oswan.maximum_menu-1) menu_oswan.Choose_Menu_value = 0;
 			break;
-			case SDLK_TAB:
-			case SDLK_LEFT:
+			case SDLK_m:
+			case SDLK_l:
 				profile_config = 0;
 			break;
-			case SDLK_BACKSPACE:
-			case SDLK_RIGHT:
+			case SDLK_n:
+			case SDLK_r:
 				profile_config = 1;
 			break;
-			case SDLK_LCTRL:
-			case SDLK_RETURN:
+			case SDLK_a:
+			case SDLK_s:
 				Set_Menu(SETTINGS_KEY_SCREEN);
 			break;
-			case SDLK_LALT:
+			case SDLK_b:
 				Set_Menu(EMULATOR_MAIN_MENU);
 			break;
 		}
diff -uprN c/main/sdl/input.c d/main/sdl/input.c
--- c/main/sdl/input.c	2025-02-04 11:17:45.302682517 +0800
+++ d/main/sdl/input.c	2025-02-04 11:16:15.992682650 +0800
@@ -23,7 +23,7 @@ void exit_button(void)
 {
 	keys = SDL_GetKeyState(NULL);
 	/* Get to Menu hotkey */
-	if ( ((keys[SDLK_ESCAPE] == SDL_PRESSED) && (keys[SDLK_RETURN] == SDL_PRESSED)) 
+	if ((keys[SDLK_q] == SDL_PRESSED) ) 
 	#ifdef RG99
 	|| keys[SDLK_PAGEUP] == SDL_PRESSED
 	#endif
