diff -uprN a/Makefile b/Makefile
--- a/Makefile	2025-12-18 20:12:50.276420380 +0800
+++ b/Makefile	2025-12-18 20:16:20.936418407 +0800
@@ -5,7 +5,7 @@
 
 TARGET = pocketsnes/PocketSNES
 
-CROSS_COMPILE := arm-linux-
+CROSS_COMPILE := /opt/FunKey-sdk-2.3.0/bin/arm-funkey-linux-gnueabihf-
 
 CC    := $(CROSS_COMPILE)gcc
 CXX   := $(CROSS_COMPILE)g++
@@ -21,7 +21,7 @@ INCLUDE = -I src \
 		-I menu -I src/linux -I src/snes9x
 
 CCFLAGS =  $(INCLUDE) -D__LINUX__ -D__DINGUX__ -DFOREVER_16_BIT  $(SDL_CFLAGS)
-CCFLAGS += -Ofast -march=armv5te -mtune=arm926ej-s -marm
+CCFLAGS += -Ofast -fdata-sections -ffunction-sections -march=armv7-a+neon-vfpv4 -mtune=cortex-a7 -mfpu=neon-vfpv4
 CCFLAGS += --fast-math -fomit-frame-pointer -fno-strength-reduce -falign-functions=2 -fno-stack-protector
 CCFLAGS += -flto=4 -fwhole-program -fuse-linker-plugin -fmerge-all-constants
 CCFLAGS += -fdata-sections -ffunction-sections
diff -uprN a/sal/linux/sal.c b/sal/linux/sal.c
--- a/sal/linux/sal.c	2025-12-18 20:14:29.656419449 +0800
+++ b/sal/linux/sal.c	2025-12-18 20:19:51.956416432 +0800
@@ -35,35 +35,35 @@ static u32 sal_Input(int held)
 	inputHeld = 0;
 
 	u8 *keystate = SDL_GetKeyState(NULL);
-	if (keystate[SDLK_LCTRL])
+	if (keystate[SDLK_b])
 		inputHeld |= SAL_INPUT_B;
-	if (keystate[SDLK_LALT])
+	if (keystate[SDLK_a])
 		inputHeld |= SAL_INPUT_A;
-	if (keystate[SDLK_SPACE])
+	if (keystate[SDLK_y])
 		inputHeld |= SAL_INPUT_Y;
-	if (keystate[SDLK_LSHIFT])
+	if (keystate[SDLK_x])
 		inputHeld |= SAL_INPUT_X;
-	if (keystate[SDLK_TAB])
+	if (keystate[SDLK_m])
 		inputHeld |= SAL_INPUT_L;
-	if (keystate[SDLK_BACKSPACE])
+	if (keystate[SDLK_n])
 		inputHeld |= SAL_INPUT_R;
-	if (keystate[SDLK_RETURN])
+	if (keystate[SDLK_s])
 		inputHeld |= SAL_INPUT_START;
-	if (keystate[SDLK_ESCAPE])
+	if (keystate[SDLK_k])
 		inputHeld |= SAL_INPUT_SELECT;
-	if (keystate[SDLK_UP])
+	if (keystate[SDLK_u])
 		inputHeld |= SAL_INPUT_UP;
-	if (keystate[SDLK_DOWN])
+	if (keystate[SDLK_d])
 		inputHeld |= SAL_INPUT_DOWN;
-	if (keystate[SDLK_LEFT])
+	if (keystate[SDLK_l])
 		inputHeld |= SAL_INPUT_LEFT;
-	if (keystate[SDLK_RIGHT])
+	if (keystate[SDLK_r])
 		inputHeld |= SAL_INPUT_RIGHT;
-	if (keystate[SDLK_RCTRL] || (keystate[SDLK_ESCAPE] && keystate[SDLK_RETURN]))
+	if (keystate[SDLK_q] || (keystate[SDLK_k] && keystate[SDLK_s]))
 		inputHeld |= SAL_INPUT_MENU;
-	if (keystate[SDLK_ESCAPE] && keystate[SDLK_TAB])
+	if (keystate[SDLK_k] && keystate[SDLK_a])
 		inputHeld |= SAL_INPUT_QUICKLOAD;
-	if (keystate[SDLK_ESCAPE] && keystate[SDLK_BACKSPACE])
+	if (keystate[SDLK_k] && keystate[SDLK_b])
 		inputHeld |= SAL_INPUT_QUICKSAVE;
 
 	mInputRepeat = inputHeld;
