diff -uprN a/Makefile.funkey b/Makefile.funkey
--- a/Makefile.funkey	2025-12-18 12:05:26.917229552 +0800
+++ b/Makefile.funkey	2025-12-18 12:02:09.227226969 +0800
@@ -1,6 +1,6 @@
-PRGNAME     = vb.elf
-CC          = /opt/funkey-toolchain/bin/arm-linux-gcc
-CXX 		= /opt/funkey-toolchain/bin/arm-linux-g++
+PRGNAME     = vb
+CC          = /opt/FunKey-sdk-2.3.0/bin/arm-linux-gcc
+CXX 		= /opt/FunKey-sdk-2.3.0/bin/arm-linux-g++
 
 #### Configuration
 
diff -uprN a/shell/input/sdl/input_emu.c b/shell/input/sdl/input_emu.c
--- a/shell/input/sdl/input_emu.c	2025-12-18 21:10:36.136387931 +0800
+++ b/shell/input/sdl/input_emu.c	2025-12-18 21:11:52.486387216 +0800
@@ -45,7 +45,7 @@ uint16_t Read_Input(void)
 			case SDL_KEYUP:
 				switch(event.key.keysym.sym)
 				{
-					case SDLK_HOME:
+					case SDLK_q:
 						emulator_state = 1;
 					break;
 					default:
@@ -64,7 +64,7 @@ uint16_t Read_Input(void)
 	}
 	
 #ifdef GKD350_BUG_INPUT
-	if (keys[SDLK_RETURN] && keys[SDLK_ESCAPE]) emulator_state = 1;
+	if (keys[SDLK_k] && keys[SDLK_s]) emulator_state = 1;
 #endif
 	
 	// UP
diff -uprN a/shell/menu/menu.c b/shell/menu/menu.c
--- a/shell/menu/menu.c	2025-12-18 11:53:05.737219869 +0800
+++ b/shell/menu/menu.c	2025-12-18 12:00:43.287225847 +0800
@@ -79,43 +79,43 @@ static void config_load()
 		fclose(fp);
 		
 		/* Fix it accordingly if previous config file is detected*/
-		if (option.config_buttons[1] == SDLK_RIGHT)
+		if (option.config_buttons[1] == SDLK_r)
 		{
 			/* Default mapping for Horizontal */
-			option.config_buttons[0] = SDLK_UP; // UP
-			option.config_buttons[1] = SDLK_DOWN; // DOWN
-			option.config_buttons[2] = SDLK_LEFT; // LEFT
-			option.config_buttons[3] = SDLK_RIGHT; // RIGHT
+			option.config_buttons[0] = SDLK_u; // UP
+			option.config_buttons[1] = SDLK_d; // DOWN
+			option.config_buttons[2] = SDLK_l; // LEFT
+			option.config_buttons[3] = SDLK_r; // RIGHT
 
 			/* Default mapping for Horizontal */
 			option.config_buttons[10] = 0; // DPAD2-UP
 			option.config_buttons[11] = 0; // DPAD2-DOWN
-			option.config_buttons[12] = SDLK_LSHIFT; // DPAD2-LEFT
-			option.config_buttons[13] = SDLK_SPACE; // DPAD2-RIGHT	
+			option.config_buttons[12] = SDLK_x; // DPAD2-LEFT
+			option.config_buttons[13] = SDLK_y; // DPAD2-RIGHT	
 		}
 	}
 	else
 	{
 		/* Default mapping for Horizontal */
-		option.config_buttons[0] = SDLK_UP; // UP
-		option.config_buttons[1] = SDLK_DOWN; // DOWN
-		option.config_buttons[2] = SDLK_LEFT; // LEFT
-		option.config_buttons[3] = SDLK_RIGHT; // RIGHT
+		option.config_buttons[0] = SDLK_u; // UP
+		option.config_buttons[1] = SDLK_d; // DOWN
+		option.config_buttons[2] = SDLK_l; // LEFT
+		option.config_buttons[3] = SDLK_r; // RIGHT
 
-		option.config_buttons[4] = SDLK_LCTRL; // A
-		option.config_buttons[5] = SDLK_LALT; // B
+		option.config_buttons[4] = SDLK_a; // A
+		option.config_buttons[5] = SDLK_b; // B
 
-		option.config_buttons[6] = SDLK_TAB; // L
-		option.config_buttons[7] = SDLK_BACKSPACE; // R
+		option.config_buttons[6] = SDLK_m; // L
+		option.config_buttons[7] = SDLK_n; // R
 
-		option.config_buttons[8] = SDLK_RETURN; // START
-		option.config_buttons[9] = SDLK_ESCAPE; // SELECT
+		option.config_buttons[8] = SDLK_s; // START
+		option.config_buttons[9] = SDLK_k; // SELECT
 		
 		/* Default mapping for Horizontal */
 		option.config_buttons[10] = 0; // DPAD2-UP
 		option.config_buttons[11] = 0; // DPAD2-DOWN
-		option.config_buttons[12] = SDLK_LSHIFT; // DPAD2-LEFT
-		option.config_buttons[13] = SDLK_SPACE; // DPAD2-RIGHT
+		option.config_buttons[12] = SDLK_x; // DPAD2-LEFT
+		option.config_buttons[13] = SDLK_y; // DPAD2-RIGHT
 
 		option.fullscreen = 1;
 	}
@@ -235,35 +235,35 @@ static void Input_Remapping()
             {
                 switch(Event.key.keysym.sym)
                 {
-                    case SDLK_UP:
+                    case SDLK_u:
                         currentselection--;
                         if (currentselection < 1)
                         {
 							currentselection = 14;
 						}
                         break;
-                    case SDLK_DOWN:
+                    case SDLK_d:
                         currentselection++;
                         if (currentselection == 15)
                         {
 							currentselection = 1;
 						}
                         break;
-                    case SDLK_LCTRL:
-                    case SDLK_RETURN:
+                    case SDLK_a:
+                    case SDLK_s:
                         pressed = 1;
 					break;
-                    case SDLK_ESCAPE:
+                    case SDLK_k:
                         option.config_buttons[currentselection - 1] = 0;
 					break;
-                    case SDLK_LALT:
+                    case SDLK_b:
 					case SDLK_p:
                         exit_input = 1;
 					break;
-                    case SDLK_LEFT:
+                    case SDLK_l:
 						if (currentselection > 8) currentselection -= 8;
 					break;
-                    case SDLK_RIGHT:
+                    case SDLK_r:
 						if (currentselection < 9) currentselection += 8;
 					break;
 					default:
@@ -288,7 +288,7 @@ static void Input_Remapping()
 						{
 							if (Event.type == SDL_KEYDOWN)
 							{
-								if (Event.key.keysym.sym != SDLK_RCTRL)
+								if (Event.key.keysym.sym != SDLK_q)
 								{
 									option.config_buttons[currentselection - 1] = Event.key.keysym.sym;
 									exit_map = 1;
@@ -462,27 +462,27 @@ void Menu()
             {
                 switch(Event.key.keysym.sym)
                 {
-                    case SDLK_UP:
+                    case SDLK_u:
                         currentselection--;
                         if (currentselection == 0)
                             currentselection = 6-SCALING_SOFTWARE_OFFSET;
                         break;
-                    case SDLK_DOWN:
+                    case SDLK_d:
                         currentselection++;
                         if (currentselection == 7-SCALING_SOFTWARE_OFFSET)
                             currentselection = 1;
                         break;
                     case SDLK_END:
-                    case SDLK_RCTRL:
-                    case SDLK_LALT:
+                    case SDLK_q:
+                    case SDLK_b:
 						pressed = 1;
 						currentselection = 1;
 						break;
-                    case SDLK_LCTRL:
-                    case SDLK_RETURN:
+                    case SDLK_a:
+                    case SDLK_s:
                         pressed = 1;
                         break;
-                    case SDLK_LEFT:
+                    case SDLK_l:
                         switch(currentselection)
                         {
                             case 2:
@@ -498,7 +498,7 @@ void Menu()
 							#endif
                         }
                         break;
-                    case SDLK_RIGHT:
+                    case SDLK_r:
                         switch(currentselection)
                         {
                             case 2:
diff -uprN a/shell/video/funkey/video_blit.c b/shell/video/funkey/video_blit.c
--- a/shell/video/funkey/video_blit.c	2025-12-18 11:52:33.077219442 +0800
+++ b/shell/video/funkey/video_blit.c	2025-12-18 11:53:29.207220175 +0800
@@ -83,7 +83,7 @@ void Set_Video_Menu()
 	 * It could be an issue with the old SDL version that we are using..
 	 * I don't know. We need to compare it against the GCW0's SDL.
 	 * */
-	sdl_screen = SDL_SetVideoMode(240, 240, BPP_TOSET, SDL_FLAGS);
+	sdl_screen = SDL_SetVideoMode(320, 240, BPP_TOSET, SDL_FLAGS);
 	
 	SDL_FillRect(sdl_screen, NULL, 0);
 	SDL_Flip(sdl_screen);
@@ -95,7 +95,7 @@ void Set_Video_Menu()
 
 void Set_Video_InGame()
 {
-	sdl_screen = SDL_SetVideoMode(240, 240, BPP_TOSET, SDL_FLAGS);
+	sdl_screen = SDL_SetVideoMode(320, 240, BPP_TOSET, SDL_FLAGS);
 	
 #if defined(WANT_32BPP)
 	internal_pix = (uint32_t*)vb_surface->pixels;
