diff -uprN a/Makefile.miyoo b/Makefile.miyoo
--- a/Makefile.miyoo	2026-02-19 08:51:42.379452420 +0800
+++ b/Makefile.miyoo	2026-02-19 08:54:43.199454070 +0800
@@ -1,6 +1,6 @@
 # OpenJazz makefile
 
-HOMEPATH = /mnt
+HOMEPATH = /mnt/FunKey
 TARGET ?= OpenJazz
 VERSION ?= $(shell date +%Y-%m-%d\ %H:%M)
 RELEASEDIR = package
@@ -11,7 +11,7 @@ DESTDIR = games
 SECTION = games
 ALIASES = aliases.txt
 
-CROSS_COMPILE=/opt/miyoo/bin/arm-linux-
+CROSS_COMPILE=/opt/mmiyoo/bin/arm-linux-gnueabihf-
 CC=$(CROSS_COMPILE)gcc
 CXX=$(CROSS_COMPILE)g++
 SYSROOT=$(shell $(CC) --print-sysroot)
diff -uprN a/build_opk.sh b/build_opk.sh
--- a/build_opk.sh	1970-01-01 08:00:00.000000000 +0800
+++ b/build_opk.sh	2026-03-08 10:22:59.981322065 +0800
@@ -0,0 +1,7 @@
+#!/bin/bash
+
+make -f Makefile.miyoo
+/opt/mmiyoo/bin/arm-linux-gnueabihf-strip OpenJazz -o _opk/openjazz
+mv dist/data _opk/
+mksquashfs _opk/* ../_opk/openjazz_funkey-s.opk -all-root -no-xattrs -noappend -no-exports
+
diff -uprN a/src/io/controls.cpp b/src/io/controls.cpp
--- a/src/io/controls.cpp	2026-02-19 08:53:16.449453279 +0800
+++ b/src/io/controls.cpp	2026-02-19 08:56:27.909455026 +0800
@@ -32,10 +32,10 @@
 
 #include "loop.h"
 
-#define DEFAULT_KEY_UP              (SDLK_UP)
-#define DEFAULT_KEY_DOWN            (SDLK_DOWN)
-#define DEFAULT_KEY_LEFT            (SDLK_LEFT)
-#define DEFAULT_KEY_RIGHT           (SDLK_RIGHT)
+#define DEFAULT_KEY_UP     (SDLK_UP)
+#define DEFAULT_KEY_DOWN   (SDLK_DOWN)
+#define DEFAULT_KEY_LEFT   (SDLK_LEFT)
+#define DEFAULT_KEY_RIGHT  (SDLK_RIGHT)
 #if defined (_WIN32)
     #define DEFAULT_KEY_JUMP            (SDLK_RALT)
     #define DEFAULT_KEY_SWIM            (SDLK_RALT)
@@ -59,16 +59,16 @@
     #define DEFAULT_KEY_YES             (SDLK_LCTRL)
     #define DEFAULT_KEY_NO              (SDLK_LALT)
 #elif defined(MIYOO) /*PocketGO mapping in description*/
-    #define DEFAULT_KEY_JUMP            (SDLK_LALT) /* A button */
-    #define DEFAULT_KEY_SWIM            (SDLK_LALT) /* A button */
-    #define DEFAULT_KEY_FIRE            (SDLK_LCTRL) /* B button */
-    #define DEFAULT_KEY_CHANGE          (SDLK_SPACE) /* Y button */
-    #define DEFAULT_KEY_ENTER           (SDLK_LALT) /* A button */
-    #define DEFAULT_KEY_ESCAPE          (SDLK_ESCAPE)  /* SELECT button */
-    #define DEFAULT_KEY_STATS           (SDLK_TAB)  /* L1 button */
-    #define DEFAULT_KEY_PAUSE           (SDLK_RETURN) /* START button */
-    #define DEFAULT_KEY_YES             (SDLK_LALT) /* A button */
-    #define DEFAULT_KEY_NO              (SDLK_LCTRL) /* B button */
+    #define DEFAULT_KEY_JUMP            (SDLK_SPACE) /* A button */
+    #define DEFAULT_KEY_SWIM            (SDLK_SPACE) /* A button */
+    #define DEFAULT_KEY_FIRE   (SDLK_LCTRL) /* B button */
+    #define DEFAULT_KEY_CHANGE          (SDLK_LALT) /* Y button */
+    #define DEFAULT_KEY_ENTER           (SDLK_SPACE) /* A button */
+    #define DEFAULT_KEY_ESCAPE          (SDLK_ESCAPE)  /* SELECT button */
+    #define DEFAULT_KEY_STATS  (SDLK_TAB)  /* L1 button */
+    #define DEFAULT_KEY_PAUSE  (SDLK_RETURN) /* START button */
+    #define DEFAULT_KEY_YES             (SDLK_SPACE) /* A button */
+    #define DEFAULT_KEY_NO     (SDLK_LCTRL) /* B button */
 #elif defined(DINGOO) 
     #define DEFAULT_KEY_JUMP            (SDLK_LCTRL)
     #define DEFAULT_KEY_SWIM            (SDLK_LCTRL)
diff -uprN a/src/menu/menu.cpp b/src/menu/menu.cpp
--- a/src/menu/menu.cpp	2026-02-19 08:53:00.069453129 +0800
+++ b/src/menu/menu.cpp	2026-02-19 08:56:58.259455303 +0800
@@ -225,14 +225,14 @@ int Menu::textInput (const char* request
 
 		}
 
-		if ((character == SDLK_DELETE) && (cursor < strlen(input))) {
+		if ((character == SDLK_RCTRL) && (cursor < strlen(input))) {
 
 			for (count = cursor; count < STRING_LENGTH; count++)
 				input[count] = input[count + 1];
 
 		}
 
-		if ((character == SDLK_BACKSPACE) && (cursor > 0)) {
+		if ((character ==SDLK_BACKSPACE) && (cursor > 0)) {
 
 			for (count = cursor - 1; count < STRING_LENGTH; count++)
 				input[count] = input[count + 1];
